<!DOCTYPE html>
<html lang="en-us">

  <head>
  <meta charset="utf-8">
  <meta name="robots" content="all,follow">
  <meta name="googlebot" content="index,follow,snippet,archive">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <title>Software</title>
  <meta name="author" content="" />
  

  
  <meta name="keywords" content="marine, data, R">
  

  
  <meta name="description" content="Site template made by Saskia using hugo">
  

  <meta name="generator" content="Hugo 0.49" />

  <link href='//fonts.googleapis.com/css?family=Roboto:400,100,100italic,300,300italic,500,700,800' rel='stylesheet' type='text/css'>

  
  <link rel="stylesheet" href="//maxcdn.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css">

  
  
  <link rel="stylesheet" href="/css/bootstrap.css" >



  
  <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.4.1/css/all.css" integrity="sha384-5sAR7xN1Nv6T6+dT2mhtzEpVJvfS3NScPQTrOxhwjIuvcA67KV2R5Jz6kr4abQsz" crossorigin="anonymous">
  

  
  
  
  <link rel="stylesheet" href="/css/github_custom.css" rel="stylesheet" id="theme-stylesheet">
  <script src="/js/highlight.pack.js"></script>
  <script>hljs.initHighlightingOnLoad();</script>


  
  
  
  


  
  <link href="/css/animate.css" rel="stylesheet">

  
  
  
    <link href="/css/style.default.css" rel="stylesheet" id="theme-stylesheet">
  


  
  <link href="/css/style.blue_custom.css" rel="stylesheet">


  
  
    <!--[if lt IE 9]>
        <script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>
        <script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
    <![endif]-->
  

  
  <link rel="shortcut icon" href="/img/favicon.ico" type="image/x-icon" />
  <link rel="apple-touch-icon" href="/img/apple-touch-icon.png" />
  

  <link href="/css/owl.carousel.css" rel="stylesheet">
  <link href="/css/owl.theme.css" rel="stylesheet">

  <link rel="alternate" href="/index.xml" type="application/rss+xml" title="Marine Data Science">

  
  <meta property="og:title" content="Software" />
  <meta property="og:type" content="website" />
  <meta property="og:url" content="/software//" />
  <meta property="og:image" content="img/mds_logo_notext@2x.png" />



</head>


  <body> 

    <div id="all">

        <header>

          <div class="navbar-affixed-top" data-spy="affix" data-offset-top="200">

    <div class="navbar navbar-default yamm" role="navigation" id="navbar">

        <div class="container">
            <div class="navbar-header">
                <a class="navbar-brand home" href="/">
                    <img src="/img/mds_logo_notext@2x.png" alt="Software logo" class="hidden-xs hidden-sm" style="max-height:50px;">
                    <img src="/img/mds_logo_notext@2x.png" alt="Software logo" class="visible-xs visible-sm" style="max-height:40px;">
                    <span class="sr-only">Software - go to homepage</span>
                </a>
                <div class="navbar-buttons">
                    <button type="button" class="navbar-toggle btn-template-main" data-toggle="collapse" data-target="#navigation">
                      <span class="sr-only">Toggle Navigation</span>
                        <i class="fa fa-align-justify"></i>
                    </button>
                </div>
            </div>
            

            <div class="navbar-collapse collapse" id="navigation">
                <ul class="nav navbar-nav navbar-right">
                  
                  <li class="dropdown">
                    
                    <a href="/">Home</a>
                    
                  </li>
                  
                  <li class="dropdown">
                    
                    <a href="/mds/">MDS</a>
                    
                  </li>
                  
                  <li class="dropdown">
                    
                    <a href="/blog/">Blog</a>
                    
                  </li>
                  
                  <li class="dropdown">
                    
                    <a href="/courses/">Courses</a>
                    
                  </li>
                  
                  <li class="dropdown">
                    
                    <a href="/data/">Data</a>
                    
                  </li>
                  
                  <li class="dropdown">
                    
                    <a href="/software/">Software</a>
                    
                  </li>
                  
                  <li class="dropdown">
                    
                    <a href="/about/">About</a>
                    
                  </li>
                  
                  <li class="dropdown">
                    
                    <a href="/contact/">Contact</a>
                    
                  </li>
                  
                </ul>
            </div>
            

            <div class="collapse clearfix" id="search">

                <form class="navbar-form" role="search">
                    <div class="input-group">
                        <input type="text" class="form-control" placeholder="Search">
                        <span class="input-group-btn">

                    <button type="submit" class="btn btn-template-main"><i class="fa fa-search"></i></button>

                </span>
                    </div>
                </form>

            </div>
            

        </div>
    </div>
    

</div>




        </header>
        

        <div id="heading-breadcrumbs">
    <div class="container">
        <div class="row">
            <div class="col-md-12">
                <h1 class="title toc-ignore">Software</h1>
            </div>
        </div>
    </div>
</div>



        <div id="content">
            

              <div class="container" id="toc_software">
  <div class="row">

      
      <div class="col-md-9"> 
        <section>
            <p><img src="../img/software/programming_languages.png"  class="w3-round" alt="Programming languages" style="max-width: 100%; height: auto; with: auto\9;"> <br></p>
<div id="which-programming-languages-should-you-learn-for-marine-data-science" class="section level2">
<h2>Which programming languages should you learn for marine data science?</h2>
<p>First of all, there is no easy answer and no consensus either. Choosing from the broad range of programming languages the right one to invest time and effort learning it the next month or years is not trivial. It all boils down to what you plan to do with it and whether you want an open source program or can afford a license.</p>
<p>If you are a marine scientist that manages rather small datasets or queries data from existing databases with specific web interfaces or user-firendly programs and then run some statistics on this data for either a report or a peer-reviewed articles, a program specifically developed for data analysis might be the best option for you. On the other hand, if you are a data scientist in a marine research institute, you might also need to build relational databases for storage of institute-specific datasets or develop tools for specific method applications, communicate the work through different channels such as software or coding repositories as well as static and dynamic websites. In that case knowing several additional programming languages such as SQL, HTML, CSS. If you work with big datasets such as sequence and high-throughput screening data or digital image data from optical sensor systems you will need a programming languages which performs particularly fast. The key here is to understand your usage requirements in terms of generality vs specificity, as well as your personal preferred development style of performance vs productivity.</p>
<p>Some noteworthy aspects when choosing a language are:</p>
<ul>
<li><strong>Purpose of application</strong>: Not every lanugage offers similar functionality for data storage, visualization, and analysis. Also, much of the day-to-day work in data science revolves around collecting and processing raw data for quick visualization and simple analysis. For this, fancy machine learning packages are not necessarily helpful.</li>
<li><strong>Compiled, interpreted, or JIT-compiled</strong> language
<ul>
<li>Compiled = good performance: A so-called <em>compiler</em> translates the language into the machine’s native language; this lets the code run fast, particularly if the compiler is effective at optimizing. But the compiling process may take time and the resulting code might not port well across operating systems.</li>
<li>Interpreted = good protability: A so-called <em>interpreter</em> reads and executes the language. The advantage here is generally a better portability across systems and little compiling time. On the other side, intepretated languages are usually slower than equivalent compiled languages.</li>
<li>JIT-compiled = balance between performance and portability: <em>Just-In-Time</em> compiled languages are quickly compiled when programs written in them need to be run, usually with little optimization.</li>
</ul></li>
<li><strong>High or low level</strong> language
<ul>
<li>High: Focuses more on concepts that are easy to understand by the human mind (objects, mathematical functions); it is, hence, easier to develop programs based on high-level languages</li>
<li>Low: More similar to the computer’s native language, i.e. machine code, and particularly suitable for high performance programs that need access to the hardware. Low level langauges are difficult to port to other platforms and practically never interpreted. However, many languages offer low-level elements.</li>
</ul></li>
<li>What <strong>type system</strong>: type systems refer to the rules that the different types of variables of a language have to follow. Different types refer to the
<ul>
<li>checking (static vs. dynamic): If a language is statically typed, then the compiler or interpreter does the type checking once before the program runs or is compiled. If the language is dynamically type checked, then the types are checked at run-time. Generally, statically typed languages are considerably more fail-proof than dynamically typed.</li>
<li>strength (strong vs weak)</li>
<li>expression (manifest vs. inferred)</li>
<li>safety</li>
</ul></li>
<li><strong>Standardization</strong>: Some languages follow a formal standard (e.g. ISO or ANSI), which can be relevant if different programs need to work together.</li>
</ul>
<p>So let’s see what other data scientiests and statisticians think, but bare in mind that some of these languages and tools do not serve the same purpose and are better used complementary.</p>
<div id="common-technologies-and-languages-in-data-science" class="section level4">
<h4>Common technologies and languages in data science</h4>
<p>Amongst the ~16,000 data scientists that participated in the 2017 <a href="https://www.kaggle.com/amberthomas/kaggle-2017-survey-results" id="kaggle_anchor">kaggle survey</a> (see also under <a href="../mds/index.html">MDS</a>) the following data science/analytics tools, technologies, and languages were used most often:</p>
<pre class="r"><code>programs_used %&gt;% with(wordcloud(selections, n, scale = c(3, 0.2), max.words = 150, 
    random.order = FALSE, rot.per = 0.35, use.r.layout = TRUE, colors = brewer.pal(6, 
        &quot;Blues&quot;)[c(4, 5, 6, 7, 8, 9)]))</code></pre>
<div class="figure" style="text-align: left"><span id="fig:unnamed-chunk-2"></span>
<img src="/software_files/figure-html/unnamed-chunk-2-1.png" alt="Wordcloud of most common languages and tools amongst all data scientist." width="768" />
<p class="caption">
Figure 1: Wordcloud of most common languages and tools amongst all data scientist.
</p>
</div>
<pre class="r"><code>p_acad &lt;- programs_acad %&gt;% ggplot(aes(x = selections, y = percent, fill = selections)) + 
    geom_col() + guides(fill = &quot;none&quot;) + colormap::scale_fill_colormap(discrete = T, 
    colormap = colormaps$salinity, reverse = T) + theme_classic() + ylim(0, 
    100) + coord_flip() + ylab(&quot;Percent&quot;) + xlab(&quot;&quot;)

p_stat &lt;- programs_stat %&gt;% ggplot(aes(x = selections, y = percent, fill = selections)) + 
    geom_col() + guides(fill = &quot;none&quot;) + colormap::scale_fill_colormap(discrete = T, 
    colormap = colormaps$salinity, reverse = T) + theme_classic() + ylim(0, 
    100) + ylab(&quot;&quot;) + coord_flip() + ylab(&quot;Percent&quot;) + xlab(&quot;&quot;)

gridExtra::grid.arrange(p_acad, p_stat, ncol = 2)</code></pre>
<div class="figure" style="text-align: left"><span id="fig:unnamed-chunk-4"></span>
<img src="/software_files/figure-html/unnamed-chunk-4-1.png" alt="Barplot of most common languages and tools amongst academic data scientist (left) and statisticians (right)." width="768" />
<p class="caption">
Figure 2: Barplot of most common languages and tools amongst academic data scientist (left) and statisticians (right).
</p>
</div>
<p><br> <!-- Obviously, the most common languages used are Python, R, and SQL and Jupyter Notebooks as tool. In academia, SQL is less important and instead Matlab often applied.  --></p>
<!-- Python was the most commonly used data analysis tool across employed data scientists overall, but more Statisticians are still loyal to R. -->
</div>
<div id="recommended-programming-language" class="section level4">
<h4>Recommended programming language</h4>
<p>Respondents were also asked what programming language they would recommend a new data scientist to learn first:</p>
<pre class="r"><code>recom_lang %&gt;% ggplot(aes(x = selections, y = percent, fill = selections)) + 
    geom_col() + guides(fill = &quot;none&quot;) + scale_fill_viridis_d(direction = 1, 
    option = &quot;inferno&quot;) + theme_classic() + ylab(&quot;Percent&quot;) + xlab(&quot;Recommended Language&quot;)</code></pre>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-6"></span>
<img src="/software_files/figure-html/unnamed-chunk-6-1.png" alt="Barplot showing the programming languages data scientist recommend most for newbies (ranked by their percentage off recommendation)." width="768" />
<p class="caption">
Figure 3: Barplot showing the programming languages data scientist recommend most for newbies (ranked by their percentage off recommendation).
</p>
</div>
<p><br> In marine sciences, the most common programming language is probably R, as it is most rich in packages and libraries related to methods applied in this field, closely followed by Matlab. However, if you plan to dive into the world of numerical modelling or you will work with big datasets such as genomics you are most likely better off with Python. I personally love R and have worked with it for many years, though at the time of my PhD it wasn’t as easy to learn - without the many online tutorials and huge archives of questions and solutions at stack overflow that nowadays exist. For my own data storage I use relational databases, so SQL is my language of choice. And to communicate results and for teaching data science I also started to learn HTML and CSS. While the integration of RMarkdown into RStudio is great for producing output as PDF, presentation, HTML page or even entire website you need to learn a bit of HTML and CSS and potentially JavaScript, Bootstrap, etc. if you want to add your personal touch to it. <br></p>
</div>
<div id="overview-of-programming-languages-and-applications" class="section level3">
<h3>Overview of programming languages and applications</h3>
<p>Here is a short summary of different programming languages and data analysis software that are more or less common in marine data science, their pros and cons and links where you can learn more about it. Since I’m fully into R I also provide you later with a detailed - but not complete- list of R packages that are useful for general or marine- specific tasks. <br> <br></p>
<p><img src="../img/software/logo_r.png" alt="logo r" style="float:left;clear:left;height:50px;border:0;padding-right:10px;"></p>
<!-- <h4 style="padding-top:20px;"><a href="https://www.r-project.org">www.r-project.org</a></h4>  -->
<div id="www.r-project.org" class="section level4">
<h4><a href="https://www.r-project.org">www.r-project.org</a></h4>
<iframe width="640" height="360" style="margin:10px;" src="https://www.youtube.com/embed/y-HOQnMsylA?feature=player_detailpage" frameborder="0" allowfullscreen>
</iframe>
<p><strong>R</strong> is a free software environment for statistical computing and graphics. It compiles and runs on a wide variety of UNIX platforms, Windows and MacOS. R was released in 1995 as a direct descendant of the older S programming language and is currently supported by the R Foundation for Statistical Computing. It is, like S, designed around a true computer language, and it allows users to add additional functionality by defining new functions. Much of the system is itself written in the R dialect of S, which makes it easy for users to follow the algorithmic choices made. For computationally-intensive tasks, C, C++ and Fortran code can be linked and called at run time. Advanced users can write C code to manipulate R objects directly.</p>
<p>In some ways R is not a typical programming language, as it’s not a general purpose one. R’s roots lies in statistics, as it has been developed specifically to deal with such needs. It provides a wide variety of statistical (linear and nonlinear modelling, classical statistical tests, time-series analysis, classification, clustering, etc.) and graphical techniques, and is highly extensible. This is one of its particular stengths, its package system that allows the functional extension of R. There are about eight packages supplied with the R distribution and many more are available through the CRAN family of Internet sites covering a very wide range of modern statistics. At present the official CRAN package repository features &gt; 10,000 packages covering common and not-so-common tasks! It is through this, and its ties with the academic community, that R keeps up to date with all latest developments in data science and machine learning. Another of R’s strengths is the ease with which well-designed publication-quality plots can be produced, including mathematical symbols and formulae where needed.</p>
<p>One of R’s downside as a result of its open-source and package system is that it has been plagued by issues such as memory management and security, and its syntax is not very straightforward or disciplined. <br></p>
<p><img src="../img/software/logo_rstudio.jpg" alt="logo rstudio" style="float:left;clear:left;height:50px;border:0;padding-right:10px;"></p>
</div>
<div id="www.rstudio.com" class="section level4">
<h4><a href="https://www.rstudio.com">www.rstudio.com</a></h4>
<p><strong>RStudio</strong> is a free and open-source integrated development environment (IDE) for R and has become the standard amongst R users in recent years. The program has an integrated text editor and data/package manager and helps writing and executing R code and analyzing data with R. It provides further version control, LaTeX integration, keyboard shortcuts, debugging tools and supports various different static and dynamic output formats through R Markdown. RStudio runs either on a desktop (Windows, Mac, and Linux) or in a web browser connected to RStudio Server. It is partly written in the C++ programming language and uses the Qt framework for its graphical user interface. The bigger percentage of the code is written in Java.</p>
<p><img src="../img/software/logo_python.png" alt="logo python" style="float:left;clear:left;height:50px;border:0;padding-right:10px;"></p>
</div>
<div id="www.python.org" class="section level4">
<h4><a href="https://www.python.org">www.python.org</a></h4>
<p><strong>Python</strong> is an interpreted high-level programming language for general-purpose programming. First released in 1991, Python has evolved to the fastest-growing programming language with an extensive range of purpose-built modules and community support. Many online services provide a Python API. Python has a design philosophy that emphasizes code readability, notably using significant whitespace, and clear programming on both small and large scales. The language features a dynamic type system and automatic memory management. It supports multiple programming paradigms, including object-oriented, imperative, functional and procedural, and has a large and comprehensive standard library. Interpreters are available for many operating systems. Python is considered as excellent all-rounder and particularly suitable for machine learning with libraries such as Google’s Tensorflow.</p>
<p>A nice online tutorial on Python is available at <a href="https://www.w3schools.com/python/default.asp">www.w3schools.com/python/</a></p>
<p><img src="../img/software/logo_cpp.png" alt="logo c++" style="float:left;clear:left;height:50px;border:0;padding-right:10px;"></p>
</div>
<div id="www.isocpp.org" class="section level4">
<h4><a href="https://isocpp.org">www.isocpp.org</a></h4>
<p><strong>C++</strong> is a statically-typed, free-form, intermediate-level general-purpose, compiled programming language, with implementations of it available on many platforms. Many vendors provide C++ compilers, including the Free Software Foundation, Microsoft, Intel, and IBM. C++ has imperative, object-oriented and generic programming features, while also providing facilities for low-level memory manipulation. It was developed by Bjarne Stroustrup at Bell Labs since 1979, as an extension of the C language -the mother of all programming languages- as he wanted an efficient and flexible language similar to C, which also provided high-level features for program organization. Since then C++ has influenced many other languages including C#, D, Java, and newer versions of C.</p>
<p>In that sense, C++ is not the ideal programming language for data analysis per se but rather for data analysis programs to base some of their functionality on (as done e.g. in R, RStudio). Code in C++ is also less readable as in R or Python. See the following example:</p>
<p>To print the text “Hello world!” you would type in C++</p>
<pre class="rcpp"><code>#Include stdout

Int main()

{

Std::cout &lt;&lt;&quot;Hello, world!&quot;;

}</code></pre>
<p>In R you would write</p>
<pre class="r"><code>print(&quot;Hello world!&quot;)</code></pre>
<p>And in Python</p>
<pre class="python"><code>Print (&quot;Hello world!&quot;)</code></pre>
<p><img src="../img/software/logo_sql.png" alt="logo sql" style="height:50px;border:0;padding-right:10px;"></p>
<p><strong>SQL</strong> (‘Structured Query Language’) defines, manages and queries relational databases. The language appeared by 1974 and has since undergone many implementations, but the core principles remain the same. Some implementations are free (e.g. MySQL, PostgreSQL), others proprietary (e.g. Oracle CHECK!!!!). SQL represents one of the favourites amongst data sciencists as indicated by the <a href="#kaggle_anchor">kaggle survey</a>. From advanced databases to NOSQL data stacks, SQL is used to parse data in the ETL stack, for advanced queries and for dealing especially with large databases, reducing the turnaround time for online requests by its fast processing time. Thus, SQL is more useful as a data processing language than as an advanced analytical tool. However, learning SQL can be a good addition into skills required for data science and ML experts, as this is looked after by most employers as a preferred skill set.</p>
<p>Common SQL database systems are: <a href="https://www.mysql.com/">MySQL</a>, <a href="https://www.microsoft.com/en-us/sql-server/">SQL Server</a>, <a href="https://mariadb.org/">MariaDB</a>, <a href="https://www.postgresql.org/">PostgreSQL</a>, <a href="https://www.oracle.com/index.html">Oracle</a>, Sybase, Informix, and MS Access.</p>
<p>A nice online tutorial on SQL is available at <a href="https://www.w3schools.com/sql/default.asp">www.w3schools.com/sql/</a></p>
<p><img src="../img/software/logo_java.png" alt="logo java" style="float:left;clear:left;height:50px;border:0;padding-right:10px;"></p>
</div>
<div id="www.java.com" class="section level4">
<h4><a href="https://www.java.com/">www.java.com</a></h4>
<p><strong>Java</strong> is another general-purpose, high-level, compiled and object-oriented language intended to have as few implementation dependencies possible and to run on all platforms without the need for recompilation. Its special feature is “write once, run anywhere” (WORA). Once the code is compiled to bytecode, it can run on any Java virtual machine (JVM) regardless of computer architecture. Thus, portability is one of the great facets of this language. Java was originally developed at Sun Microsystems, now owned by Oracle, and is nowadays one of the most popular programming languages amongst client-server web applications. A large number of companies, especially big multinational companies use the language to develop backend systems and desktop apps. Beside the criticism of speed<a href="#fn1" class="footnoteRef" id="fnref1"><sup>1</sup></a> and safety<a href="#fn2" class="footnoteRef" id="fnref2"><sup>2</sup></a>, Jave is commonly used in software engineering. In marine data science, howeever it is less common.</p>
<p>A nice online tutorial on Java is available at <a href="https://www.w3schools.com/java/default.asp">www.w3schools.com/java/</a></p>
<p><img src="../img/software/logo_fortran.jpg" alt="logo fortran" style="float:left;clear:left;height:50px;border:0;padding-right:10px;"></p>
</div>
<div id="www.fortran.com" class="section level4">
<h4><a href="http://www.fortran.com/">www.fortran.com</a></h4>
<p><strong>Fortran</strong> (formerly FORTRAN, derived from FORmula TRANslation) is a general-purpose, compiled imperative programming language that was designed for scientists and engineer in the 1950s by IBM. Fortran came to dominate this area of programming early on and has been in continuous use for over half a century in computationally intensive areas such as numerical weather prediction, finite element analysis, computational fluid dynamics, computational physics, crystallography and computational chemistry. It is a popular language for high-performance computing[4] and is used for programs that benchmark and rank the world’s fastest supercomputers.[5] Revision of the language led to FORTRAN 77, the language we use today. The standard for FORTRAN 90 is now available although not yet in widespread use. F77 is a subset of F90.</p>
<p><img src="../img/software/logo_matlab.jpg" alt="logo matlab" style="float:left;clear:left;height:50px;border:0;padding-right:10px;"></p>
</div>
<div id="www.mathworks.com" class="section level4">
<h4><a href="https://www.mathworks.com/">www.mathworks.com</a></h4>
<p><strong>MATLAB</strong> is a numerical computing language developed by Mathworks in the 1980s and based of C, C++, and Java programming languages. MATLAB is a fast and stable program with solid algorithms for complex math and used throughout academia and industry. MATLAB is well-suited for quantitative applications with sophisticated mathematical requirements such as signal processing, Fourier transforms, matrix algebra and image processing. It also has some great in-built plotting options.</p>
<p>The major drawback of MATLAB are its high pricing for basic licenses and all additional libraries. Also, since MATLAB is not an open source language and, hence, does not allow users to easily extend the functionality, MATLAB does not keep up with latest developments in data science and machine learning as R or Python.</p>
<p><img src="../img/software/logo_sas.png" alt="logo sas" style="float:left;clear:left;height:40px;border:0;padding-right:10px;"></p>
</div>
<div id="www.sas.com" class="section level4">
<h4><a href="https://www.sas.com">www.sas.com</a></h4>
<p><strong>SAS</strong> is one of the most popular languages amongst statistians but less so in the data science community. It has a wide range of statistical functions with a user friendly GUI that helps data scientists learn quickly. It is an easy to learn programming language and preferred as a must have language for beginners entering analytics industry. The two biggest drawbacks are - similar to Matlab - its high pricing for the licenses as well as the limited flexibility for extending its functionality.</p>
<p><img src="../img/software/logo_jupyter.png" alt="logo jupyter" style="float:left;clear:left;height:50px;border:0;padding-right:10px;"></p>
</div>
<div id="httpjupyter.org" class="section level4">
<h4><a href="http://jupyter.org/">http://jupyter.org</a></h4>
<p>The <strong>Jupyter Notebook</strong> is an open-source web application that allows you to create and share documents that contain live code, equations, visualizations and narrative text. The Notebook has support for over 40 programming languages, including Python, R, Julia, and Scala. Notebooks can be shared with others using email, Dropbox, GitHub and the Jupyter Notebook Viewer. The code run in Jypyter can produce rich, interactive output: HTML, images, videos, LaTeX, and custom MIME types. <br></p>
</div>
<div id="other" class="section level4">
<h4>Other</h4>
<p><strong>Perl</strong> (<a href="https://www.perl.org">www.perl.org</a>)<br> Perl is an open source, general-purpose, interpreted cross platform programming language originally developed for text manipulation and now used for a wide range of tasks including system administration, web development, network programming, and GUI development.</p>
<p><strong>Julia</strong> (<a href="https://julialang.org/">www.julialang.org</a>)<br> Julia is a high level dynamic programming language designed to address the needs of high performance numerical analysis and computational science and is increasingly gaining popularity amongst data scientists. The base library written in Julia combines also open source C and Fortran libraries for linear algebra, random number generation, signal processing, and string processing.</p>
<p><strong>Scala</strong> (<a href="https://www.scala-lang.org/">www.scala-lang.org</a>)<br> Scala is a general-purpose, scalable language, which is organized to run on the Java Virtual Machine (JVM). Scale has a large user interface, is quite user-friendly and can be used to write apps.</p>
<p><strong>Ruby</strong> (<a href="https://www.ruby-lang.org/en/">www.ruby-lang.org</a>)<br> Ruby is a dynamic, interpreted, reflective, object-oriented, general-purpose programming language developed in the 1990s that has a more coherent design than its close rivals (Pyhton and Perl). While the Ruby-on-Rails framework initially gained a big popularity in Web applications, it did not maintain momentum due to performance and scalability problems. <br></p>
</div>
</div>
</div>
<div id="r-packages" class="section level1">
<h1>R packages</h1>
<p>The R package system has grown so much and fast (&gt; 10,000 package in 2018) that it is shear impossible to list all the useful package a marine (data) scientist would like to use at some point. So please use the following list rather as a starting point. You can find also great overviews of useful packages at the CRAN task views for specific topics:</p>
<ul>
<li>For an overview of packages for ecological and environmental analyses in R see the <a href="https://cran.r-project.org/web/views/Environmetrics.html">Environmetrics</a> CRAN task view maintained by Gavin Simpson.</li>
<li>For time series analysis see the <a href="https://cran.r-project.org/web/views/TimeSeries.html">TimeSeries</a> CRAN task view maintained by Rob J Hyndman.</li>
<li>For spatial and spatio-temporal analyses see the <a href="https://cran.r-project.org/web/views/Spatial.html">Spatial</a> CRAN task view maintained by Roger Bivand as well as the <a href="https://cran.r-project.org/web/views/SpatioTemporal.html">SpatioTemporal</a> task view maintained by Edzer Pebesma.</li>
<li>For machine learning and statistical learning see <a href="https://cran.r-project.org/web/views/MachineLearning.html">MachineLearning</a> mainatined by Torsten Hothorn.</li>
<li>For multivariate analyses see the <a href="https://cran.r-project.org/web/views/Multivariate.html">Multivariate</a> maintained by Paul Hewson and the <a href="https://cran.r-project.org/web/views/Cluster.html">Cluster</a> CRAN task view maintained by Friedrich Leisch and Bettina Gruen</li>
</ul>
<p>To stay up-to-date check out also <a href="https://www.r-bloggers.com/">R-bloggers</a>, which is a blog with news on latest advances, package presentations and online tutorials.</p>
<div id="accessingretrieving-data" class="section level2">
<h2>Accessing/retrieving data</h2>
<ul>
<li><a href="https://github.com/hadley/rvest">rvest</a>: A general package to easily harvest (scrape) data from websites.</li>
<li><a href="http://www.rforge.net/DATRAS/Tutorial.html">DATRAS</a>: Read and manipulate trawl survey data from the <a href="http://ices.dk/marine-data/data-portals/datras">DATRAS</a> database.</li>
<li><a href="https://github.com/seananderson/ramlegacy">ramlegacy</a>: Package to download, import, convert, and cache the <a href="http://ramlegacy.org/">RAM Legacy Stock Assessment Database</a>.</li>
<li><a href="https://github.com/ropensci/rnoaa">rnoaa</a>: Access a whole bunch of NOAA datasets → for more info see also the <a href="https://ropensci.org/tutorials/rnoaa_tutorial/">online tutorial</a>.</li>
<li><a href="https://github.com/iobis/robis">robis</a>: R client for the Ocean Biogeographic Information System (<a href="http://www.iobis.org">OBIS</a>) API.</li>
</ul>
<p>The non-profit initative <a href="https://ropensci.org">rOpenSci</a> aims to make scientific data retrieval reproducible and has developed in the past an ecosystem of open source tools and reviewed community developed software. Various R packages have been built within the rOpenSci community useful for marine data scientists:</p>
<ul>
<li><a href="https://github.com/ropensci/pangaear">pangaear</a>: This package offers tools to interact with the <a href="https://www.pangaea.de">PANGAEA</a> Database, including functions for searching for data, fetching datasets by dataset ID, and working with the PANGAEA OAI-PMH service.</li>
<li>Climate data:
<ul>
<li><a href="https://github.com/ropensci/ccafs">ccafs</a>: R client for Climate Change, Agriculture, and Food Security (CCAFS) General Circulation Models (GCM) data.</li>
</ul></li>
<li>Geospatial data:
<ul>
<li><a href="https://github.com/ropensci/mregions">mregions</a>: Package to access all of the data at <a href="http://www.marineregions.org">www.marineregions.org</a>. Allows visualization of marine regions with associated data paired with analysis, use of marine region geospatial boundaries to query data providers (e.g., <a href="http://www.iobis.org">OBIS</a>), and geocoding.</li>
<li><a href="https://github.com/ropensci/rnaturalearth">rnaturalearth</a>: An R package to hold and facilitate interaction with natural earth map data. The <a href="www.naturalearthdata.com">Natural Earth</a> website structures vector data by scale, category and type. These determine the filenames of downloads. ‘rnaturalearth’ uses this structure to facilitate download (like an API).</li>
</ul></li>
<li>Taxonomic data (see also this <a href="https://github.com/ropensci/taxonomy">overview</a>):
<ul>
<li><a href="https://github.com/ropensci/Reol">Reol</a>: R client for all data from Encyclopedia of Life (<a href="http://eol.org/">EOL</a>), which includes taxonomic data. Note that ‘taxize’ also interfaces with EOL data, only the taxonomic data though.</li>
<li><a href="https://github.com/ropensci/ritis">ritis</a>: R client for the Integrated Taxonomic Information System (<a href="https://www.itis.gov">ITIS</a>) database of taxonomic data.</li>
<li><a href="https://github.com/ropensci/taxize">taxize</a>: This is the place to go for most taxononomic data. ‘taxize’ connects to a lot of data sources, and has consistent data outputs across the data sources.</li>
<li><a href="https://github.com/ropensci/taxizedb">taxizedb</a>: Evolved from the ‘taxize’ package to handle larger taxonomic data sets or infrequent internet access. ‘taxizedb’ downloads taxonomic database dumps from many different providers and allows either SQL queries or data wrangling with the ‘dplyr’ package.</li>
<li><a href="https://github.com/ropensci/taxizesoap">taxizesoap</a>: An extension to ‘taxize’, but for data sources that use SOAP data transfer protocol, which is hard to support in R. This package is not on CRAN.</li>
<li><a href="https://github.com/ropensci/worrms">worrms</a>: R client for the World Register of Marine Species (<a href="https://github.com/ropensci/worrms">WoRMS</a>) API.</li>
</ul></li>
<li>Species data (see also <a href="https://cran.rstudio.com/web/packages/spocc/vignettes/spocc_vignette.html">here</a> for an overview):
<ul>
<li><a href="https://github.com/ropensci/bold">bold</a>: This package retrieves data from the <a href="http://boldsystems.org">BOLD</a> database of barcode clusters, and allows for searching of over 1.7M public records using multiple search criteria including sequence data, specimen data, specimen plus sequence data, as well as trace files.</li>
<li><a href="https://github.com/ropensci/originr">originr</a>: Provides functions to scrapes pecies origin data from the web (e.g. from ITIS, Encyclopedia of Life, <a href="http://www.iucngisd.org/gisd/">Global Invasive Species Database</a>, <a href="http://www.griis.org">Global Register of Introduced and Invasive Species</a>).</li>
<li><a href="https://github.com/ropensci/rredlist">rredlist</a>: R client for the <a href="https://www.iucnredlist.org">IUCN Red List</a>.</li>
<li><a href="https://github.com/ropensci/rfishbase">rfishbase</a>: R interface to the <a href="https://www.fishbase.org">fishbase</a> database. ‘rfishbase 3.0’ queries pre-compressed tables from a static server and employs local caching (through memorization) to provide much greater performance and stability, particularly for dealing with large queries involving 10s of thousands of species.</li>
<li><a href="https://github.com/ropensci/rgbif">rgbif</a>: Interface to the Global Biodiversity Information Facility API → for more info see <a href="https://github.com/ropensci/rgbif" class="uri">https://github.com/ropensci/rgbif</a>.</li>
<li><a href="https://github.com/ropensci/traits">traits</a>: R client for various sources of species trait data.</li>
</ul></li>
<li>Fisheries data:
<ul>
<li><a href="https://github.com/ropensci/rfisheries">rfisheries</a>: Package for interacting with fisheries databases at <a href="www.openfisheries.org">openfisheries.org</a>.</li>
</ul></li>
</ul>
</div>
<div id="data-wrangling-and-processing" class="section level2">
<h2>Data wrangling and processing</h2>
<ul>
<li><a href="https://github.com/tidyverse/tidyverse">tidyverse</a>: A collection of R packages that share an underlying design philosophy, grammar, and data structures. The tidyverse package is designed to make it easy to install and load core packages from the tidyverse in a single command (these includes the packages:
<ul>
<li><a href="http://readr.tidyverse.org/">readr</a> for data import,</li>
<li><a href="http://tidyr.tidyverse.org/">tidyr</a> for data tidying,</li>
<li><a href="http://tibble.tidyverse.org/">tibble</a> for tibbles as modified data frames,</li>
<li><a href="http://dplyr.tidyverse.org/">dplyr</a> for data manipulation,</li>
<li><a href="https://github.com/hadley/forcats">forcats</a> for factors,</li>
<li><a href="https://github.com/tidyverse/stringr">stringr</a> for strings,</li>
<li><a href="http://purrr.tidyverse.org/">purrr</a> for functional programming, and</li>
<li><a href="http://ggplot2.tidyverse.org/">ggplot2</a> for data visualisation)</li>
</ul></li>
<li>To manipulate/parse taxonomic names
<ul>
<li><a href="https://cran.r-project.org/web/packages/metacoder/index.html">metacoder</a>: Package to parse, manipulate, and visualize metabarcoding/taxonomic data see also the article by <a href="https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1005404">Foster et al. 2017</a></li>
<li><a href="https://github.com/ropensci/taxview">taxview</a>: Package to summarize taxonomic data in the form of getting taxonomic hierarchy data (via the ‘taxize’ package) to facilitate taxonomic summaries.</li>
</ul></li>
<li>NetCDF (Network Common Data Form): Most climate data (observations, model output, reanalysis) is in the NetCDF format, a self-documenting, machine-independent format for creating and distributing arrays of gridded data:
<ul>
<li><a href="https://cran.r-project.org/web/packages/ncdf4/index.html">ncdf4</a>: Provides a high-level R interface to import netCDF files (version 4 or earlier), which are binary data files that are portable across platforms and include metadata information in addition to the data sets. It is also easy to create new netCDF dimensions, variables, and files, in either version 3 or 4 format, and manipulate existing netCDF files. This package replaces the former ‘ncdf’ package, which only worked with netcdf version 3 files.</li>
<li><a href="https://github.com/hypertidy/tidync">tidync</a>: Systematic approaches to NetCDF data extraction, manipulation and visualization. The data extracted can be used directly as an array, or in “long form” form as a data frame for “tidy” analysis and visualization contexts.</li>
</ul></li>
<li><a href="https://github.com/AustralianAntarcticDivision/EchoviewR">EchoviewR</a>: An R package for automated processing of active acoustic data using Echoview → for more info see also the article by <a href="https://www.frontiersin.org/articles/10.3389/fmars.2015.00015/full">Harrison et al., 2015</a></li>
</ul>
</div>
<div id="general-visualisation-packages" class="section level2">
<h2>General visualisation packages</h2>
<p>Visualisation packages belong to one of two grammar styles: the old school base graphics and the <a href="http://ggplot2.tidyverse.org/">ggplot2</a> system for declaratively creating graphics, based on The Grammar of Graphics, which has gained a large popularity in recent years:</p>
<div id="base-graphics" class="section level4">
<h4>base graphics:</h4>
<ul>
<li><a href="https://cran.r-project.org/web/packages/diagram/index.html">diagram</a>: Visualises simple graphs (networks) based on a transition matrix, utilities to plot flow diagrams, visualising webs, electrical networks, etc.</li>
<li><a href="https://cran.r-project.org/web/packages/mapplots/index.html">mapplots</a>: Creates simple maps, adds sub-plots like pie plots to a map or any other plot, and formats, plots and exports gridded data. The package was originally developed for displaying fisheries data but most functions can be used for more generic data visualisation.</li>
<li><a href="https://cran.r-project.org/web/packages/shape/index.html">shape</a>: Provides functions for plotting graphical shapes such as ellipses, circles, cylinders, arrows, etc.</li>
</ul>
</div>
<div id="ggplot2-based-graphics" class="section level4">
<h4>ggplot2-based graphics:</h4>
<p>An overview for ggplot2 extensions (40 in Nov. 2018) can be found here: <a href="https://www.ggplot2-exts.org/" class="uri">https://www.ggplot2-exts.org/</a></p>
<ul>
<li><a href="https://cran.r-project.org/web/packages/cowplot/vignettes/introduction.html">cowplot</a>: Provides a publication-ready theme for ‘ggplot2’, which requires a minimum amount of fiddling with sizes of axis labels, plot backgrounds, etc.</li>
<li><a href="https://github.com/thomasp85/gganimate">gganimate</a>: Wraps the animation package to create animated ggplot2 plots.</li>
<li><a href="https://github.com/thomasp85/ggraph">ggraph</a>: Supports relational data structures such as networks, graphs, and trees.</li>
<li><a href="https://cran.r-project.org/web/packages/ggmuller/index.html">ggmuller</a>: Creates Muller plots for visualizing evolutionary dynamics.<br />
</li>
<li><a href="https://cran.r-project.org/web/packages/ggridges/vignettes/introduction.html">ggridges</a>: Creates ridgeline plots that are partially overlapping line plots to create the impression of a mountain range. They can be quite useful for visualizing changes in distributions over time or space.</li>
<li><a href="https://github.com/jrnold/ggthemes">ggthemes</a>: Some extra geoms, scales, and themes for ggplot<br />
</li>
<li><a href="https://github.com/aloy/qqplotr">qqplotr</a>: Provides functions to plot quantile-quantile (Q-Q) and probability-probability (P-P) points, lines, and confidence bands.</li>
<li><a href="https://cran.r-project.org/web/packages/ggvis/index.html">ggvis</a>: An implementation of an interactive grammar of graphics, taking the best parts of ‘ggplot2’, combining them with the reactive framework of ‘shiny’ and drawing web graphics using ‘vega’ → for more info see also the <a href="https://ggvis.rstudio.com/">ggvis website</a>.</li>
</ul>
</div>
<div id="color-schemes" class="section level4">
<h4>Color schemes:</h4>
<ul>
<li><a href="https://cran.r-project.org/web/packages/colormap/index.html">colormap</a>: Allows to generate colors from palettes defined in the colormap module of ‘Node.js’ → see <a href="https://github.com/bpostlethwaite/colormap" class="uri">https://github.com/bpostlethwaite/colormap</a> for more information. In total it provides 44 distinct palettes made from sequential and/or diverging colors. In addition to the pre defined palettes you can also specify your own set of colors. There are also scale functions that can be used with ‘ggplot2’.</li>
<li><a href="https://cran.r-project.org/web/packages/colourpicker/index.html">colourpicker</a>: A colour picker that can be used as an input in Shiny apps or Rmarkdown documents. The colour picker supports alpha opacity, custom colour palettes, and many more options. A Plot Colour Helper tool is available as an RStudio Addin, which helps you pick colours to use in your plots.</li>
<li><a href="https://cran.r-project.org/web/packages/RColorBrewer/index.html">RColorBrewer</a>: For visualization, the colour palettes provided in this package are very useful, and may be modified or extended using the <code>colorRampPalette()</code> function provided with R → for more info see also <a href="http://colorbrewer2.org" class="uri">http://colorbrewer2.org</a>.</li>
</ul>
</div>
</div>
<div id="time-series" class="section level2">
<h2>Time series</h2>
<p>See also the <a href="https://cran.r-project.org/web/views/TimeSeries.html">TimeSeries</a> CRAN task view for complementary information.</p>
<p>The base package <strong>‘stats’</strong> offers various function for time series data: <code>ts()</code> creates time-series objects from vectors; the <code>ar()</code> and <code>arima()</code> functions fit autoregressive (AR), moving average (MA), autoregressive moving average (ARMA) and integrated ARMA (ARIMA) time series model to the data. Classes “POSIXct” and “POSIXlt” implement the POSIX standard for date/time (intra-day) information and also support time zones and daylight savings time. However, the time zone computations require some care and might be system-dependent. Internally, “POSIXct” objects are the number of seconds since 1970-01-01 00:00:00 GMT. The ‘tidyverse’ package <a href="https://lubridate.tidyverse.org/">lubridate</a> provides functions that facilitate certain POSIX-based computations.</p>
<ul>
<li><a href="https://cran.r-project.org/package=TSA">TSA</a>: Contains R functions and datasets detailed in the book “<a href="https://www.springer.com/de/book/9780387759586">Time Series Analysis with Applications in R</a> by Jonathan Cryer and Kung-Sik Chan.</li>
<li><a href="https://cran.r-project.org/package=tseries">tseries</a> and <a href="https://cran.r-project.org/package=zoo">zoo</a>: Both packages provide general handling and analysis of time series data. Irregular time series can be handled by using package ‘zoo’ as well as the function <code>irts()</code> in ‘tseries’ → for more info on ‘zoo’ see also the <a href="https://cran.r-project.org/web/packages/zoo/vignettes/zoo-quickref.pdf">Quick Reference vignette</a>.</li>
<li><a href="https://cran.r-project.org/package=tsibble">tsibble</a> and <a href="https://cran.r-project.org/package=tibbletime">tibbletime</a> build on the tidy principle and extend the tibble to store and manage temporal data. ‘tsibble’ provides a ‘tbl_ts’ class (the ‘tsibble’), which is built on top of the ‘tibble’. The ‘tsibble’ aims at easily manipulating and analysing temporal data, including counting and filling time gaps, aggregate over calendar periods, performing rolling window calculations, and etc. ‘tibbletime’ creates time aware tibbles that allows time-based subsetting on tibbles, a quick summary and aggregation of results by time periods, and creating columns that can be used as ‘dplyr’ time-based groups.</li>
<li><a href="https://cran.r-project.org/package=xts">xts</a>: eXtensible Time Series - Provide for uniform handling of R’s different time-based data classes by extending zoo, maximizing native format information preservation and allowing for user level customization and extension, while simplifying cross-class interoperability.</li>
</ul>
<div id="time-series-modelling" class="section level4">
<h4>Time series modelling</h4>
<ul>
<li><a href="https://cran.r-project.org/package=CommonTrend">CommonTrend</a>: Provides tools to extract and plot common trends from a cointegration system using different approaches that are widely used in macroeconomics.</li>
<li><a href="https://cran.r-project.org/package=dse">dse</a>: This package provides a variety of more advanced estimation methods and tools for multivariate, linear, time-invariant, time series models → for more info see also the <a href="https://cran.r-project.org/web/packages/dse/vignettes/Guide.pdf">User Guide</a>.</li>
<li><a href="https://cran.r-project.org/package=dynlm">dynlm</a>: Convenient interface to fitting time series regressions via ordinary least squares.</li>
<li><a href="https://cran.r-project.org/package=dyn">dyn</a>: Provides a different approach to that of ‘dynlm’ by allowing time series data to be used with any regression function written in the style of lm such as <code>lm()</code>, <code>glm()</code>, <code>loess()</code>, <code>quantreg::rq()</code>, <code>MASS::rlm()</code>, <code>MCMCpack::MCMCregress()</code>, <code>quantreg::rq()</code>, <code>randomForest::randomForest()</code> amongst others, whilst preserving the time series information.</li>
<li><a href="https://cran.r-project.org/package=forecast">forecast</a>: Provides methods and tools for displaying and analysing univariate time series forecasts including exponential smoothing via state space models and automatic ARIMA modelling.<br />
</li>
<li><a href="https://cran.r-project.org/package=MARSS">MARSS</a>: Provides maximum-likelihood parameter estimation for constrained and unconstrained linear Multivariate Autoregressive State-Space (MARSS) models fit to multivariate time-series data. Fitting is primarily via an Expectation-Maximization (EM) algorithm, although fitting via the BFGS algorithm (using the optim function) is also provided. MARSS models are a class of dynamic linear model (DLM) and vector autoregressive model (VAR) model. Functions are provided for parametric and innovations bootstrapping, Kalman filtering and smoothing, bootstrap model selection criteria (AICb), confidences intervals via the Hessian approximation and via bootstrapping and calculation of auxiliary residuals for detecting outliers and shocks → for more information see also the <a href="https://cran.r-project.org/web/packages/MARSS/vignettes/UserGuide.pdf">User Guide</a>, and the package website <a href="https://nwfsc-timeseries.github.io/MARSS/" class="uri">https://nwfsc-timeseries.github.io/MARSS/</a>.<br />
</li>
<li><a href="https://cran.r-project.org/package=pastecs">pastecs</a>: Regularisation, decomposition and analysis of space-time series.</li>
<li><a href="https://cran.r-project.org/package=tsDyn">tsDyn</a>: Implements nonlinear autoregressive (AR) time series models. For univariate series, a non-parametric approach is available through additive nonlinear AR. Parametric modeling and testing for regime switching dynamics is available when the transition is either direct (TAR: threshold AR) or smooth (STAR: smooth transition AR, LSTAR). For multivariate series, one can estimate a range of TVAR or threshold cointegration TVECM models with two or three regimes → see also this <a href="https://cran.r-project.org/web/packages/tsDyn/vignettes/tsDyn.pdf">Introduction</a>.</li>
</ul>
</div>
<div id="frequency-analysis" class="section level4">
<h4>Frequency analysis</h4>
<ul>
<li>Spectral density estimation is provided by <code>spectrum()</code> in the ‘stats’ package, including the periodogram, smoothed periodogram and AR estimates.</li>
<li>The <a href="https://cran.r-project.org/package=forecast">forecast</a> package also provides some simple harmonic regression facilities via the <code>fourier()</code> function.</li>
<li><a href="https://cran.r-project.org/package=wavelets">wavelets</a>: Includes computing wavelet filters, wavelet transforms and multiresolution analyses.</li>
<li><a href="https://cran.r-project.org/package=WaveletComp">WaveletComp</a>: Provides some tools for wavelet-based analysis of univariate and bivariate time series including cross-wavelets, phase-difference (with filtering options), significance with simulation algorithms.</li>
</ul>
</div>
<div id="changepoint-detection" class="section level4">
<h4>Changepoint detection</h4>
<ul>
<li>Univariate time series
<ul>
<li><a href="https://cran.r-project.org/package=bcp">bcp</a>: Bayesian analysis of change points in the mean - provides an implementation of the Barry and Hartigan (1993) product partition model for the normal errors change point problem using Markov Chain Monte Carlo. It also extends the methodology to regression models on a connected graph (Wang and Emerson, 2015); this allows estimation of change point models with multivariate responses → for more info see also the article by <a href="https://www.jstatsoft.org/article/view/v023i03/v23i03.pdf">Erdman &amp; Emerson (2007)</a>.</li>
<li><a href="https://cran.r-project.org/package=changepoint">changepoint</a>: Implements various mainstream and specialised changepoint methods for finding single and multiple changepoints within data. Many popular non-parametric and frequentist methods are included → for more info see this useful <a href="http://members.cbio.mines-paristech.fr/~thocking/change-tutorial/RK-CptWorkshop.html">tutorial</a> or the article by <a href="https://www.jstatsoft.org/article/view/v058i03/v58i03.pdf">Killick &amp; Eckley (2014)</a>.</li>
<li><a href="https://cran.r-project.org/package=segmented">segmented</a>: Given a regression model, ‘segmented’ updates the model by adding one or more segmented (i.e., piece-wise linear) relationships. Several variables with multiple breakpoints are allowed.</li>
<li><a href="https://cran.r-project.org/package=strucchange">strucchange</a>: Detects (multiple) structural changes in (linear) regression models. strucchange features tests/methods from the generalized fluctuation test framework as well as from the F test (Chow test) framework. The breakpoints in regression models with structural changes can be estimated together with confidence intervals.</li>
</ul></li>
<li>Multivariate time series
<ul>
<li><a href="https://cran.r-project.org/package=changepoint.mv">changepoint.mv</a>: Detects the most recent changepoints for panel data consisting of many related univariate timeseries.</li>
<li><a href="https://cran.r-project.org/package=ecp">ecp</a>: Implements various non-parametric procedures for finding multiple changepoints. Two methods make use of dynamic programming and pruning, with no distributional assumptions other than the existence of certain absolute moments in one method. Hierarchical and exact search methods are included → for more info see <a href="https://cran.r-project.org/web/packages/ecp/vignettes/ecp.pdf">James &amp; Matteson (2014)</a>.</li>
<li><a href="https://cran.r-project.org/package=rioja">rioja</a>: the function <code>chclust()</code> allows the detection of changepoints in multivariate data via a constrained clustering algorithm in which clusters are constrained by sample order (e.g. ordered in space in time).</li>
</ul></li>
</ul>
</div>
</div>
<div id="spatial-analysis" class="section level2">
<h2>Spatial analysis</h2>
<p>While base R includes many functions that can be used for reading, visualising, and analysing spatial data, specific packages are required for “geographical” spatial data, where observations can be identified with geographical locations, and where additional information about these locations may be retrieved. A large number of these packages provide spatial statistical methods or interfaces to GIS, and many of them provide specific classes and plotting methods for spatial data. The following is only a small selection of most important packages. For a much more comprehensive overview of spatial packages in R see the <a href="https://cran.r-project.org/web/views/Spatial.html">Spatial</a> CRAN task view maintained by Roger Bivand as well as the <a href="https://cran.r-project.org/web/views/SpatioTemporal.html">SpatioTemporal</a> task view maintained by Edzer Pebesma.</p>
<div id="fundamental-packages" class="section level4">
<h4>Fundamental packages</h4>
<p>Packages for class definition, import/export and handling of spatial data:</p>
<ul>
<li><a href="https://cran.r-project.org/web/packages/gdistance/index.html">gdistance</a>: Provides calculation of distances and routes on geographic grids.</li>
<li><a href="https://cran.r-project.org/web/packages/geosphere/index.html">geosphere</a>: Permits computations of distance and area to be carried out on spatial data in geographical coordinates.</li>
<li><a href="https://cran.r-project.org/web/packages/landsat/index.html">landsat</a>: Provides tools for processing of Landsat or other multispectral satellite imagery. Includes relative normalization, image-based radiometric correction, and topographic correction options.</li>
<li><a href="">maptools</a>: Provides various mapping functions. It includes binary access to ‘GSHHG’ shoreline files. The package also provides interface wrappers for exchanging spatial objects with packages such as ‘PBSmapping’, ‘spatstat’, ‘maps’, ‘RArcInfo’, ‘Stata tmap’, ‘WinBUGS’, ‘Mondrian’, and others.</li>
<li><a href="https://cran.r-project.org/web/packages/raster/index.html">raster</a>: The package is a major extension of spatial data classes to virtualise access to large rasters, permitting large objects to be analysed, and extending the analytical tools available for both raster and vector data.</li>
<li><a href="https://cran.r-project.org/web/packages/rgdal/index.html">rgdal</a>: R’s interface to the popular C/C++ spatial data processing library <a href="https://www.gdal.org/">GDAL</a>. Maps may be vector-based or raster-based. The ‘rgdal’ package provides bindings to GDAL-supported raster formats and OGR-supported vector formats. It contains functions to write raster files in supported formats. The package also provides PROJ.4 projection support for vector objects ( this site provides searchable online PROJ.4 representations of projections). Both ‘GDAL’ and ‘PROJ.4’ libraries are external to the package, and, when installing the package from source, must be correctly installed first. Also the GeoJSON format can be written and read using ‘rgdal’.</li>
<li><a href="https://cran.r-project.org/web/packages/rgeos/index.html">rgeos</a>: R’s interface to the powerful vector processing library <a href="http://trac.osgeo.org/geos/">GEOS</a>.</li>
<li><a href="https://cran.r-project.org/web/packages/sf/index.html">sf</a>: ‘sf’ is a new R package, which is meant to supersede ‘sp’. It provides native support for <a href="https://cran.r-project.org/web/packages/sf/vignettes/sf1.html">Simple Features</a> in R, can perform topological operations by interfacing with <a href="http://trac.osgeo.org/geos/">GEOS</a>, can read and write to a wide variety of spatial file formats courtesy of <a href="https://www.gdal.org/">GDAL</a>, and binds to <a href="https://proj4.org/">PROJ</a> for projection conversions and datum transformations. Hence, ‘sf’ appears to do everything ‘sp’, ‘rgdal’, and ‘rgeos’ did, but in a more modern and intuitive fashion. My favourite thing about ‘sf’ is that the authors have adopted the <strong>tidy data</strong> principle:
<ul>
<li>the package puts geometries in a list-column of class sfc , where each list element is a feature geometry of class sfg,</li>
<li>all functions begin with <code>st_</code> for easy RStudio tab completion,</li>
<li>snake_case is used throughout the package,</li>
<li>‘dplyr’ and ‘tidyr’ verbs have been defined for the sf objects,</li>
<li>functions are pipe-friendly,</li>
<li>AND ‘ggplot2’ is able to plot sf objects directly → for more information check out the <a href="http://r-spatial.github.io/sf/">sf website</a>.</li>
</ul></li>
<li><a href="https://cran.r-project.org/web/packages/sp/index.html">sp</a>: ‘sp’ has been the standard for storing spatial data in R and many other packages have become dependent on the sp classes. The package provides classes and methods for spatial data (points, lines, polygons, grids) and utility functions for e.g. plotting data as maps, spatial selection, retrieving coordinates, subsetting, printint, summary, etc. → you’ll find more info on the R-Forge <a href="http://rspatial.r-forge.r-project.org/">rspatial project</a> website, including a visualisation gallery.</li>
<li><a href="https://cran.r-project.org/web/packages/spacetime/index.html">spacetime</a>: Extends the shared classes defined in ‘sp’ for spatio-temporal data (see <a href="https://www.jstatsoft.org/article/view/v051i07">Spatio-Temporal Data in R</a>).</li>
<li>An alternative approach to the ‘sp’ classes and methods are the PBS packages:
<ul>
<li><a href="https://cran.r-project.org/web/packages/PBSmapping/index.html">PBSmapping</a>: Two-dimensional plotting features similar to those commonly available in a Geographic Information System. The package has evolved from fisheries research conducted at the Pacific Biological Station (PBS) in Nanaimo, British Columbia, Canada.</li>
<li><a href="https://cran.r-project.org/web/packages/PBSmodelling/index.html">PBSmodelling</a>: Provides modelling support. It focuses particularly on tools that make it easy to construct and edit a customized graphical user interface (‘GUI’).</li>
</ul></li>
</ul>
</div>
<div id="visualisation-of-spatial-data" class="section level4">
<h4>Visualisation of spatial data</h4>
<ul>
<li><a href="https://cran.r-project.org/web/packages/cartography/index.html">cartography</a>: This package allows various cartographic representations such as proportional symbols, choropleth, typology, flows or discontinuities → see also the <a href="http://riatelab.github.io/cartography/vignettes/cheatsheet/cartography_cheatsheet.pdf">cheat sheet</a>.</li>
<li><a href="https://cran.r-project.org/web/packages/ggmap/index.html">ggmap</a>: Extends the plotting package ‘ggplot2’ for maps with a collection of functions to visualize spatial data and models on top of static maps from various online sources (e.g Google Maps and Stamen Maps).</li>
<li><a href="https://cran.r-project.org/web/packages/ggsn/index.html">ggsn</a>: Adds north symbols (18 options) and scale bars in kilometers to maps in geographic or metric coordinates created with ‘ggplot2’ or ‘ggmap’.<br />
</li>
<li><a href="https://cran.r-project.org/web/packages/leafletR/index.html">leafletR</a>: Displays your spatial data on interactive web-maps using the open-source JavaScript library <a href="Leaflet.">Leaflet</a>. ‘leafletR’ can combine vector data and online map tiles from different sources.</li>
<li><a href="https://cran.r-project.org/web/packages/mapview/index.html">mapview</a>: Creates interactive visualisations of spatial data with or without background maps. Attributes of displayed features are fully queryable via pop-up windows. Additional functionality includes methods to visualise true- and false-color raster images, bounding boxes, small multiples and 3D raster data cubes.</li>
<li><a href="https://cran.r-project.org/web/packages/plotKML/index.html">plotKML</a>: Visualization of spatial and spatio-temporal objects in Google Earth. Writes sp-class, spacetime-class, raster-class and similar spatial and spatio-temporal objects to KML following some basic cartographic rules.</li>
<li><a href="https://cran.r-project.org/web/packages/rasterVis/index.html">rasterVis</a>: Used with the ‘raster’ package it can provide enhanced visualisation and interaction. It implements visualization methods for quantitative data and categorical data, both for univariate and multivariate rasters. It also provides methods to display spatio-temporal rasters and vector fields. See the website for examples → see the <a href="https://oscarperpinan.github.io/rastervis/">website</a> for more information.</li>
<li><a href="https://cran.r-project.org/web/packages/rworldmap/index.html">rworldmap</a>: Enables mapping of country level and gridded user datasets. Modern country boundaries are provided at 2 resolutions by rworldmap along with functions to join and map tabular data referenced by country names or codes. Chloropleth and bubble maps are supported and general functions to work on user supplied maps (see <a href="https://journal.r-project.org/archive/2011-1/RJournal_2011-1_South.pdf">A New R package for Mapping Global Data</a>.</li>
<li><a href="https://cran.r-project.org/web/packages/tmap/index.html">tmap</a>: Thematic maps are geographical maps in which spatial data distributions are visualized. This package offers a flexible, layer-based, and easy to use approach based on the Grammar of Graphics syntax to create thematic maps, such as choropleths and bubble maps → a nice introduction is provided <a href="https://cran.r-project.org/web/packages/tmap/vignettes/tmap-getstarted.html">here</a>!</li>
</ul>
</div>
<div id="spatial-modelling-geostatistics" class="section level4">
<h4>Spatial modelling / Geostatistics:</h4>
<p>The choice of package for spatial analysis and modelling will depend greatly on the type of data and its distribution. If the data are characterised by point support and the spatial process is continuous, geostatistical methods may be used or functions in the <a href="https://cran.r-project.org/web/packages/nlme/index.html">nlme</a> package, which is one of the base packages.</p>
<ul>
<li><a href="https://cran.r-project.org/web/packages/automap/index.html">automap</a>: This package performs an automatic interpolation by automatically estimating the variogram and then calling gstat.</li>
<li><a href="https://cran.r-project.org/web/packages/geoR/index.html">geoR</a>: Useful for model-based geostatistics including traditional, likelihood-based and Bayesian methods.</li>
<li><a href="https://cran.r-project.org/web/packages/geoRglm/index.html">geoRglm</a>: An extension to the package ‘geoR’, which must be installed first. ‘geoRglm’ provides functions for inference in generalised linear spatial models. The posterior and predictive inference is based on MCMC methods.</li>
<li><a href="https://cran.r-project.org/web/packages/gstat/index.html">gstat</a>: The package provides a wide range of functions for univariate and multivariate geostatistics, also for larger datasets. It covers, for instance, variogram modelling and plotting, various kriging methods, or sequential Gaussian or indicator (co)simulation → see also this <a href="https://cran.r-project.org/web/packages/gstat/vignettes/gstat.pdf">brief tutorial</a>.</li>
<li><a href="https://cran.r-project.org/web/packages/ngspatial/index.html">ngspatial</a>: Provides tools for analyzing ecological spatial data, especially non-Gaussian areal data.</li>
<li><a href="https://cran.r-project.org/web/packages/rtop/index.html">rtop</a>: A package for geostatistical interpolation of data with irregular spatial support such as runoff related data or data from administrative units.</li>
<li><a href="https://cran.r-project.org/web/packages/spatial/index.html">spatial</a>: This package is shipped with base R, and contains several core functions for kriging and point pattern analysis.</li>
<li><a href="https://cran.r-project.org/web/packages/SpatialTools/index.html">SpatialTools</a>: This package has an emphasis on kriging, and provides functions for prediction and simulation. It is intended to be relatively straightforward, fast, and flexible.</li>
<li><a href="https://cran.r-project.org/web/packages/spdep/index.html">spdep</a>: If the support is areal, and the spatial process is not being treated as continuous, functions provided in the ‘spdep’ package may be used. This package can also be seen as providing spatial econometrics functions, and, as noted above, provides basic functions for building neighbour lists and spatial weights, tests for spatial autocorrelation for areal data like Moran’s I, and functions for fitting spatial regression models. It provides the full range of local indicators of spatial association, such as local Moran’s I and diagnostic tools for fitted linear models, including Lagrange Multiplier tests. Spatial regression models that can be fitted using maximum likelihood include spatial lag models, spatial error models, and spatial Durbin models. For larger data sets, sparse matrix techniques can be used for maximum likelihood fits, while spatial two stage least squares and generalised method of moments estimators are an alternative.</li>
<li><a href="https://cran.r-project.org/web/packages/vardiag/index.html">vardiag</a>: Interactive variogram diagnostics may be carried out with this package.</li>
</ul>
</div>
</div>
<div id="oceanographic-analysis" class="section level2">
<h2>Oceanographic analysis</h2>
<ul>
<li><a href="https://cran.r-project.org/web/packages/AquaEnv/index.html">AquaEnv</a>: Toolbox for the experimental aquatic chemist, focused on acidification and CO2 air-water exchange.</li>
<li><a href="https://cran.r-project.org/web/packages/EnvStats/index.html">EnvStats</a>: Graphical and statistical analyses of environmental data, with focus on analyzing chemical concentrations and physical parameters, usually in the context of mandated environmental monitoring. Major environmental statistical methods found in the literature and regulatory guidance documents, with extensive help that explains what these methods do, how to use them, and where to find them in the literature. Numerous built-in data sets from regulatory guidance documents and environmental statistics literature. Includes scripts reproducing analyses presented in the book “<a href="http://www.springer.com/book/9781461484554">EnvStats: An R Package for Environmental Statistics</a>” (Millard, 2013).</li>
<li><a href="https://cran.r-project.org/web/packages/hydroGOF/index.html">hydroGOF</a>: Useful for implementing both statistical and graphical goodness-of-fit measures between observed and simulated values, mainly oriented to be used during the calibration, validation, and application of hydrological/environmental models.</li>
<li><a href="https://cran.r-project.org/web/packages/marelac/index.html">marelac</a>: Datasets, constants, conversion factors, and utilities for ‘MArine’, ‘Riverine’, ‘Estuarine’, ‘LAcustrine’ and ‘Coastal’ science. Allows to calculate diffusion constants, gas solubilities, gas exchange coefficients, the properties of seawater</li>
<li><a href="https://cran.r-project.org/web/packages/marmap/index.html">marmap</a>: A package for importing, plotting and analyzing bathymetric and topographic data. ‘marmap’ can query the ETOPO1 bathymetry and topography database hosted by the NOAA, use simple latitude-longitude-depth data in ascii format, and take advantage of the advanced plotting tools available in R to build publication-quality bathymetric maps → for more info see <a href="https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0073051">Pante &amp; Simon-Bouhet (2013)</a></li>
<li><a href="https://cran.r-project.org/web/packages/oce/index.html">oce</a>: Package to read, summarize and plot oceanographic data, including ADCP measurements, measurements made with argo floats, CTD measurements, sectional data, sea-level time series, coastline and topographic data, etc. Provides specialized functions for calculating seawater properties such as potential temperature in either the UNESCO or TEOS-10 equation of state. Produces graphical displays that conform to the conventions of the oceanographic literature → for more info see also the <a href="https://cran.r-project.org/web/packages/oce/vignettes/oce.html">vignette</a>.</li>
<li><a href="https://cran.r-project.org/web/packages/oceanmap/index.html">oceanmap</a>: Plotting toolbox for 2D oceanographic data (satellite data, sst, chla, ocean fronts &amp; bathymetry).</li>
<li><a href="https://cran.r-project.org/web/packages/OceanView/index.html">Oceanview</a>: Specifically designed for visualising complex multidimensional oceanographic data. It can produce a.o. quiver plots, vector plots, etc. in base R graphics → more infos <a href="http://www.rforscience.com/rpackages/visualisation/oceanview/">here</a> and see the <a href="https://cran.r-project.org/web/packages/OceanView/vignettes/OceanView.pdf">vignette</a>. The package is derived from 2 other visualiation packages that are also worth exploring, i.e. <a href="https://cran.r-project.org/web/packages/plot3D/index.html">plot3d</a> and <a href="https://cran.r-project.org/web/packages/plot3Drgl/index.html">plot3drgl</a>.</li>
<li><a href="https://cran.r-project.org/web/packages/seacarb/index.html">seaCarb</a>: Calculates parameters of the seawater carbonate system and assists the design of ocean acidification perturbation experiments.</li>
</ul>
</div>
<div id="fishy-packages-for-analysis" class="section level2">
<h2>Fishy packages for analysis</h2>
<p>For packages on capture-recapture analyses see also the comprehensive overview on Derek Ogle’s website <a href="http://derekogle.com/fishR/packages">fishR</a>.</p>
<ul>
<li><a href="http://www.flr-project.org/">FLR</a>: The Fisheries Library in R is a collection of tools for quantitative fisheries science, developed in the R language, that facilitates the construction of bio-economic simulation models of fisheries systems. The library includes several packages, which can be installed all together, including their dependecies, using the code <code>source(&quot;http://flr-project.org/R/instFLR.R&quot;)</code> or by selecting your choice of package from the <a href="https://github.com/flr/R">FLR github repository</a> using the code <code>install.packages(pkgs = package_name, repos=&quot;http://flr-project.org/R&quot;)</code>. A good starting point to explore FLR is <a href="http://www.flr-project.org/doc/A_quick_introduction_to_FLR.html">A quick introduction to FLR</a>.</li>
</ul>
<div id="stock-assessments-and-population-dynamic-models" class="section level4">
<h4>Stock assessments and population dynamic models</h4>
<ul>
<li><a href="https://cran.r-project.org/web/packages/CatDyn/index.html">CatDyn</a>: Useful for fishery stock assessments by Generalized Depletion Models based on fishery catch dynamics instead of fish population dynamics.</li>
<li><a href="https://cran.r-project.org/web/packages/DLMtool/index.html">DLMtool</a>: Data-Limited Methods Toolkit – Development, simultion testing, and implementation of management procedures for data-limited fisheries → see also the article by <a href="https://besjournals.onlinelibrary.wiley.com/doi/epdf/10.1111/2041-210X.12791">Mildenberger et al. (2017)</a>.</li>
<li><a href="https://github.com/marchtaylor/fishdynr">fishdynr</a>: An R package of fisheries science related population dynamics models.</li>
<li><a href="https://cran.r-project.org/web/packages/fishmethods/index.html">fishmethods</a>: Provides functions for various fisheries stock assessment methods.</li>
<li><a href="https://cran.r-project.org/web/packages/fishMod/index.html">fishMod</a>: Fits models to catch and effort data. Single-species models are 1) delta log-normal, 2) Tweedie, or 3) Poisson-gamma (G)LMs.</li>
<li><a href="https://cran.r-project.org/web/packages/FSA/index.html">FSA</a>: A variety of simple fish stock assessment methods → for more info see <a href="http://derekogle.com/FSA/" class="uri">http://derekogle.com/FSA/</a>.</li>
<li><a href="https://cran.r-project.org/web/packages/LBSPR/index.html">LBSPR</a>: Simulate expected equilibrium length composition, yield-per-recruit (YPR), and the spawning potential ratio (SPR) using the length-based SPR (LBSPR) model. Fit the LBSPR model to length data to estimate selectivity, relative apical fishing mortality, and the spawning potential ratio for data-limited fisheries.</li>
<li><a href="https://cran.r-project.org/web/packages/mixdist/index.html">mixdist</a>: Fits finite mixture distribution models to grouped data and conditional data by maximum likelihood, possibly separating age-classes from a length frequency.</li>
<li><a href="https://cran.r-project.org/web/packages/r4ss/index.html">r4ss</a>: A collection of R functions for use with Stock Synthesis, a fisheries stock assessment modeling platform written in <a href="http://www.admb-project.org/">ADMB</a> at NOAA.</li>
<li><a href="https://cran.r-project.org/web/packages/scape/index.html">scape</a>: Import, plot, and diagnose results from statistical catch-at-age models, used in fisheries stock assessment; generally used to connect with ADMB.</li>
<li><a href="https://cran.r-project.org/web/packages/ss3sim/index.html">ss3sim</a>: Framework for fisheries stock assessment simulation testing with Stock Synthesis 3.<br />
</li>
<li><a href="https://cran.r-project.org/web/packages/TropFishR/index.html">TropFishR</a>: Tropical Fisheries Analysis with R – collection of fisheries models based on the the FAO Manual “Introduction to tropical fish stock assessment”. Focus is the analysis of length-frequency data and data poor fisheries.</li>
</ul>
</div>
<div id="fishery-analyses" class="section level4">
<h4>Fishery analyses</h4>
<ul>
<li><a href="https://cran.r-project.org/web/packages/fecR/index.html">fecR</a>: Calculates fishing effort following the DG MARE Ad-Hoc Workshops on Transversal Variables in Zagreb (2015) and Nicosia (2016) → for more info see the <a href="https://cran.r-project.org/web/packages/fecR/vignettes/calculating_fishing_effort.pdf">vignette</a>.</li>
<li><a href="https://cran.r-project.org/web/packages/vmsbase/index.html">VMSBase</a>: Tool to process, analyze, combine, integrate and plot fishery data from the vessel monitoring system, automatic information system or other tracking devices, as well as the catches or landings dataset from logbooks or vessel register.</li>
</ul>
</div>
<div id="otholith-and-age-analyses" class="section level4">
<h4>Otholith and age analyses</h4>
<ul>
<li><a href="https://cran.r-project.org/web/packages/ALKr/index.html">ALKr</a>: Provides functions that implement several algorithms for generating age-length keys for fish populations from incomplete data.</li>
<li><a href="https://cran.r-project.org/web/packages/RFishBC/index.html">RFishBC</a>: Helps fisheries scientists collect growth data from calcified structures and back-calculate estimated lengths at previous ages → for more infos see also <a href="http://derekogle.com/RFishBC/" class="uri">http://derekogle.com/RFishBC/</a>.</li>
<li><a href="https://cran.r-project.org/web/packages/shapeR/index.html">shapeR</a>: The package extends previously described software used for otolith shape analysis by allowing the user to automatically extract closed contour outlines from a large number of images, perform smoothing to eliminate pixel noise, choose from conducting either a Fourier or wavelet transform to the outlines and visualize the mean shape.</li>
</ul>
</div>
<div id="others" class="section level4">
<h4>Others</h4>
<ul>
<li><a href="https://bioenergetics4.shinyapps.io/bioenergetics4/">Bioenergetics 4.0</a>: A shiny app for bioenergetics modeling.</li>
<li><a href="https://cran.r-project.org/web/packages/fishmove/index.html">fishmove</a>: Provides functions to predict fish movement parameters plotting leptokurtic fish dispersal kernels.</li>
<li><a href="https://cran.r-project.org/web/packages/RchivalTag/index.html">RchivalTag</a>: A set of functions to generate, access and analyze standard data products from archival tagging data.</li>
<li><a href="https://cran.r-project.org/web/packages/swfscMisc/index.html">swfscMisc</a>: Collection of conversion, analytical, geodesic, mapping, and plotting functions developed by the Southwest Fisheries Science Center of NOAA.</li>
</ul>
</div>
</div>
<div id="modelling-species-responses-and-other-data-supervised-learning" class="section level2">
<h2>Modelling species responses and other data (supervised learning)</h2>
<p>Analysing species response curves or modeling other data often involves the fitting of standard statistical models to ecological data and includes simple (multiple) regression, Generalised Linear Models (GLM), extended regression (e.g. Generalised Least Squares [GLS]), Generalised Additive Models (GAM), (Non-)Linear Mixed Effects Models, and classification or tree-based models.</p>
<p>The base installation of R provides <code>lm()</code> and <code>glm()</code> for fitting linear and generalised linear models, respectively, while the <code>nls()</code> function can be used for determining the nonlinear (weighted) least-squares estimates of the parameters of a nonlinear model. See also the <a href="https://cran.r-project.org/web/views/MachineLearning.html">MachineLearning</a> CRAN task view for information on supervised statistical and machine learning packages.</p>
<ul>
<li><a href="https://CRAN.R-project.org/package=akima">akima</a>: Several cubic spline interpolation methods of H. Akima for irregular and regular gridded data are available through this package, both for the bivariate and univariate case.</li>
<li><a href="https://CRAN.R-project.org/package=boot">boot</a>: Provides bootstrap functions (originally by Angelo Canty for S).</li>
<li><a href="https://cran.r-project.org/web/packages/caret/index.html">caret</a>: The caret package (short for Classification And REgression Training) contains functions to streamline the model training process for complex regression, classification and other machine learning problems. The package utilizes a number of R packages but tries not to load them all at package start-up. The package contains tools for data splitting, pre-processing, feature selection, model tuning using resampling, variable importance estimation, etc. → for more see this <a href="https://cran.r-project.org/web/packages/caret/vignettes/caret.html">Introduction</a> and the main help pages at <a href="https://topepo.github.io/caret/" class="uri">https://topepo.github.io/caret/</a>. See also this useful <a href="https://resources.rstudio.com/rstudio-cheatsheets/caret-cheat-sheet">caret cheat sheet</a>.</li>
<li><a href="https://cran.r-project.org/package=ROCR">ROCR</a>: Provides ROC graphs, sensitivity/specificity curves, lift charts, and precision/recall plots are popular examples of trade-off visualizations for specific pairs of performance measures. ‘ROCR’ is a flexible tool for creating cutoff-parameterized 2D performance curves by freely combining two from over 25 performance measures (new performance measures can be added using a standard interface). Curves from different cross-validation or bootstrapping runs can be averaged by different methods, and standard deviations, standard errors or box plots can be used to visualize the variability across the runs. ‘ROCR’ is easy to use, with only three commands and reasonable default values for all optional parameters.<br />
</li>
<li><a href="https://cran.r-project.org/web/packages/simecol/index.html">simecol</a>: ‘Simulation of ecological systems’ is a lightweight R package intended to give users an interactive environment to implement, distribute, simulate and document basic and advanced ecological models without the need to write long simulation programs. An object oriented approach is used to provide a consistent but still flexible and extensible way to implement simulation models of different types, e.g. ordinary differential equation (ODE) models, non-spatial individual-based models, grid-oriented individual-based models, particle diffusion-type / random walk models and more → for more info see the <a href="http://simecol.r-forge.r-project.org/">simecol website</a>.</li>
<li>Sensitivity analysis
<ul>
<li><a href="https://cran.r-project.org/web/packages/sensitivity/index.html">sensitivity</a>: Contains a collection of functions for factor screening, global sensitivity analysis and reliability sensitivity analysis of model output.</li>
<li><a href="https://cran.r-project.org/web/packages/fast/index.html">fast</a>: This package implements the Fourier Amplitude Sensitivity Test (FAST), a method to determine global sensitivities of a model on parameter changes with relatively few model runs.</li>
</ul></li>
</ul>
<div id="linear-non-linear-and-additive-methods-for-regression-and-classification" class="section level4">
<h4>Linear, non-linear and additive methods for regression and classification</h4>
<ul>
<li><a href="https://CRAN.R-project.org/package=car">car</a>: Provides various functions for applied regression.</li>
<li><a href="https://CRAN.R-project.org/package=lme4">class</a>: Various functions for classification, including k-Nearest Neighbor, Learning Vector Quantization and Self-Organizing Maps.</li>
<li><a href="https://CRAN.R-project.org/package=gam">gam</a>: This is an alternative package to the more popular ‘mgcv’, which provides an implementation of the S-PLUS function <code>gam()</code> that includes LOESS smooths and other Functions for fitting and working with generalized additive models, as described in chapter 7 of “Statistical Models in S” (Chambers and Hastie (eds), 1991), and “<a href="https://www.crcpress.com/Generalized-Additive-Models/Hastie-Tibshirani/p/book/9780412343902">Generalized Additive Models</a>” (Hastie and Tibshirani, 1990).</li>
<li><a href="https://CRAN.R-project.org/package=glmnet">glmnet</a>: Glmnet is a package that fits a Generalized Linear Model via penalized maximum likelihood. Offers extremely efficient procedures for fitting the entire lasso or elastic-net regularization path for linear regression, logistic and multinomial regression models, Poisson regression and the Cox model → see also this <a href="https://web.stanford.edu/~hastie/glmnet/glmnet_alpha.html">Introduction</a>.</li>
<li><a href="https://CRAN.R-project.org/package=grplasso">grplasso</a>: Fits user-specified (GLM-) models with group lasso penalty.</li>
<li><a href="https://CRAN.R-project.org/package=kknn">kknn</a>: Package for weighted k-Nearest Neighbors for Classification, Regression and Clustering.</li>
<li><a href="https://CRAN.R-project.org/package=leaps">leaps</a>: Regression subset selection, including exhaustive search.</li>
<li><a href="https://CRAN.R-project.org/package=lme4">lme4</a>: An updated approach to mixed effects models, which also fits Generalised Linear Mixed Models (GLMM) and Generalised non-Linear Mixed Models (GNLMM). The core computational algorithms are implemented using the ‘Eigen’ C++ library for numerical linear algebra and ‘RcppEigen’ “glue” → for more infos see also <a href="https://www.jstatsoft.org/article/view/v067i01/">Bates et al. (2015)</a>.</li>
<li><a href="https://CRAN.R-project.org/package=MASS">MASS</a>: A comprehensive package shipped with base R that provides a vast set of functions and datasets to support Venables and Ripley, “Modern Applied Statistics with S” (2002).</li>
<li><a href="https://CRAN.R-project.org/package=mgcv">mgcv</a>: ‘mgcv’ is the recommended package to fit Generalised Additive (Mixed) Models (GAMM), some of their extensions and other generalized ridge regressions with multiple smoothing parameter estimation by (Restricted) Marginal Likelihood, generalised cross-validation and similar. Includes a wide variety of smoothers, JAGS support and distributions beyond the exponential family.</li>
<li><a href="https://CRAN.R-project.org/package=nlme">nlme</a>: Provides functions for fitting GLS and Llinear and non-linear mixed effects models that extend the simple regression model to account for clustering, heterogeneity and correlations within the sample of observations. The package is supported by Pinheiro &amp; Bates (2000) <a href="https://link.springer.com/book/10.1007/b98882">Mixed-effects Models in S and S-PLUS</a>, Springer, New York.</li>
<li><a href="https://CRAN.R-project.org/package=pls">pls</a>: Provides the multivariate regression methods Partial Least Squares Regression (PLSR), Principal Component Regression (PCR) and Canonical Powered Partial Least Squares (CPPLS).</li>
</ul>
</div>
<div id="tree-based-models" class="section level4">
<h4>Tree-based models</h4>
<p>Tree-based models are being increasingly used in ecology, particularly for their ability to fit flexible models to complex data sets and the simple, intuitive output of the tree structure. Ensemble methods such as bagging, boosting and random forests are advocated for improving predictions from tree-based models and to provide information on uncertainty in regression models or classifiers.</p>
<ul>
<li>Recursive Partitioning :
<ul>
<li><a href="https://CRAN.R-project.org/package=maptree">maptree</a>: Provides functions with example data for graphing, pruning, and mapping models from hierarchical clustering, and classification and regression trees.</li>
<li><a href="https://CRAN.R-project.org/package=rpart">rpart</a>: This package comes with base R and provides tree-structured models for regression, classification and survival analysis, following the ideas in the <a href="https://www.crcpress.com/Classification-and-Regression-Trees/Breiman-Friedman-Stone-Olshen/p/book/9780412048418">CART</a> book → for more see this <a href="https://cran.r-project.org/web/packages/rpart/vignettes/longintro.pdf">Introduction</a>.</li>
<li><a href="https://CRAN.R-project.org/package=party">party</a>: A computational toolbox for recursive partitioning. Provides an implementation of conditional inference trees which embed tree-structured regression models into a well defined theory of conditional inference procedures. This non-parametric class of regression trees is applicable to all kinds of regression problems, including nominal, ordinal, numeric, censored as well as multivariate response variables and arbitrary measurement scales of the covariates.</li>
<li><a href="https://CRAN.R-project.org/package=REEMtree">REEMtree</a>: This package estimates regression trees with random effects as a way to use data mining techniques to describe longitudinal or panel data.</li>
<li><a href="https://CRAN.R-project.org/package=RPMM">RPMM</a>: Recursively Partitioned Mixture Model for Beta and Gaussian Mixtures - This is a model-based clustering algorithm that returns a hierarchy of classes, similar to hierarchical clustering, but also similar to finite mixture models.</li>
<li><a href="https://CRAN.R-project.org/package=tree">tree</a>: Another package recommended for computing <a href="https://www.crcpress.com/Classification-and-Regression-Trees/Breiman-Friedman-Stone-Olshen/p/book/9780412048418">CART</a>-like trees.</li>
</ul></li>
<li>Random Forest, Bagging, Boosting and Gradient Descent
<ul>
<li><a href="https://CRAN.R-project.org/package=GAMBoost">GAMBoost</a>: This package provides routines for fitting generalized linear and and generalized additive models by likelihood based boosting, using penalized B-splines.</li>
<li><a href="https://CRAN.R-project.org/package=gbm">gbm</a>: An implementation of extensions to Freund and Schapire’s AdaBoost algorithm and Friedman’s gradient boosting machine. Includes regression methods for least squares, absolute loss, t-distribution loss, quantile regression, logistic, multinomial logistic, Poisson, etc.</li>
<li><a href="https://CRAN.R-project.org/package=ipred">ipred</a>: Improved predictive models by indirect classification and bagging for classification, regression and survival problems as well as resampling based estimators of prediction error.</li>
<li><a href="https://CRAN.R-project.org/package=quantregForest">quantregForest</a>: Implements Quantile Regression Forests, a tree-based ensemble method for estimation of conditional quantiles. It is particularly well suited for high-dimensional data. Predictor variables of mixed classes can be handled.</li>
<li><a href="https://CRAN.R-project.org/package=randomForest">randomForest</a>: The Random Forest method of <a href="https://link.springer.com/article/10.1023%2FA%3A1010933404324">Breiman and Cutler (2001)</a> is implemented in ‘randomForest’, providing classification and regression based on a forest of trees using random inputs.</li>
<li><a href="https://CRAN.R-project.org/package=xgboost">xgboost</a>: R interface to Extreme Gradient Boosting, which is an efficient implementation of the gradient boosting framework from <a href="https://www.kdd.org/kdd2016/papers/files/rfp0697-chenAemb.pdf">Chen and Guestrin (2016)</a>. The package includes efficient linear model solver and tree learning algorithms. The package can automatically do parallel computation on a single machine which could be more than 10 times faster than existing gradient boosting packages. It supports various objective functions, including regression, classification and ranking. The package is made to be extensible, so that users are also allowed to define their own objectives easily.</li>
</ul></li>
<li>Bayesian
<ul>
<li><a href="https://cran.r-project.org/package=BayesTree">BayesTree</a>: An implementation of <a href="http://rob-mcculloch.org/code/BART-7-05.pdf">BART:Bayesian Additive Regression Trees</a>, by Chipman et al. (2010), where the final model is defined in terms of the sum over many weak learners (not unlike ensemble methods).</li>
</ul></li>
</ul>
</div>
<div id="support-vector-machines-and-kernel-method" class="section level4">
<h4>Support Vector Machines and Kernel Method</h4>
<ul>
<li><a href="https://cran.r-project.org/package=e1071">e1071</a>: Misc Functions of the Department of Statistics, Probability Theory Group (Formerly: E1071) - The package provides functions for latent class analysis, short time Fourier transform, fuzzy clustering, support vector machines, shortest path computation, bagged clustering, naive Bayes classifier, etc.</li>
<li><a href="https://cran.r-project.org/package=kernlab">kernlab</a>: Kernel-based machine learning methods for classification, regression, clustering, novelty detection, quantile regression and dimensionality reduction. Among other methods ‘kernlab’ includes Support Vector Machines, Spectral Clustering, Kernel PCA, Gaussian Processes and a QP solver.</li>
</ul>
</div>
<div id="neural-networks-and-deep-learning-frameworks" class="section level4">
<h4>Neural Networks and Deep Learning frameworks:</h4>
<ul>
<li>Single-hidden-layer neural network are implemented in package <a href="https://cran.r-project.org/package=nnet">nnet</a> (shipped with base R).</li>
<li><a href="https://CRAN.R-project.org/package=deepnet">deepnet</a>: Implement some deep learning architectures and neural network algorithms, including BP,RBM,DBN,Deep autoencoder, etc.</li>
<li><a href="https://cran.r-project.org/package=h2o">h2o</a>: R Interface for <a href="https://www.h2o.ai/">H2O</a>, the scalable open source machine learning platform that offers parallelized implementations of many supervised and unsupervised machine learning algorithms such as Generalized Linear Models, Gradient Boosting Machines (including XGBoost), Random Forests, Deep Neural Networks (Deep Learning), Stacked Ensembles, Naive Bayes, Cox Proportional Hazards, K-Means, PCA, Word2Vec, as well as a fully automatic machine learning algorithm (AutoML).</li>
<li><a href="https://cran.r-project.org/package=keras">keras</a>: R interface to <a href="https://keras.io">Keras</a>, a high-level neural networks ‘API’. ‘Keras’ was developed with a focus on enabling fast experimentation, supports both convolution based networks and recurrent networks (as well as combinations of the two), and runs seamlessly on both ‘CPU’ and ‘GPU’ devices → for more information see this <a href="https://cran.r-project.org/web/packages/keras/vignettes/why_use_keras.html">Introduction</a>. See also this useful <a href="https://resources.rstudio.com/rstudio-cheatsheets/keras-cheat-sheet">keras cheat sheet</a>.</li>
<li><a href="https://CRAN.R-project.org/package=rnn">rnn</a>: Implementation of a Recurrent Neural Network in R.</li>
<li><a href="https://CRAN.R-project.org/package=tensorflow">tensorflow</a>: R interface to Google’s <a href="https://www.tensorflow.org/">TensorFlow</a>, an open source software library for numerical computation using data flow graphs. Nodes in the graph represent mathematical operations, while the graph edges represent the multidimensional data arrays (tensors) communicated between them. The flexible architecture allows you to deploy computation to one or more ‘CPUs’ or ‘GPUs’ in a desktop, server, or mobile device with a single ‘API’. ‘TensorFlow’ was originally developed by researchers and engineers working on the Google Brain Team within Google’s Machine Intelligence research organization for the purposes of conducting machine learning and deep neural networks research, but the system is general enough to be applicable in a wide variety of other domains as well.</li>
</ul>
</div>
<div id="others-1" class="section level4">
<h4>Others</h4>
<ul>
<li><a href="https://CRAN.R-project.org/package=cusp">cusp</a>: Cobb’s maximum likelihood method for cusp-catastrophe modeling. Includes a <code>cusp()</code> function for model fitting, and several utility functions for plotting, and for comparing the model to linear regression and logistic curve models.</li>
<li><a href="https://CRAN.R-project.org/package=rEDM">rEDM</a>: A new implementation of Empirical Dynamic Modeling (EDM) algorithms for time series based on research software previously developed for internal use in the Sugihara Lab (UCSD/SIO). Contains C++ compiled objects that use time delay embedding to perform state-space reconstruction and nonlinear forecasting and an R interface to those objects using ‘Rcpp’. It supports the simplex projection method, S-map algorithm, convergent cross mapping and multiview embedding → see also this <a href="https://cran.r-project.org/web/packages/rEDM/vignettes/rEDM-tutorial.html">Introduction</a>.</li>
<li><a href="https://CRAN.R-project.org/package=sem">sem</a>: Functions for fitting general linear Structural Equation Models (with observed and latent variables) using the RAM approach, and for fitting structural equations in observed-variable models by two-stage least squares.</li>
</ul>
</div>
</div>
<div id="multivariate-analyses-supervised-and-unsupervised" class="section level2">
<h2>Multivariate analyses (supervised and unsupervised)</h2>
<p>R and add-on packages provide a wide range of multivariate methods, many of which are specialised ordination and clustering techniques well suited for the analysis of species data. Two packages have become the leading add-on packages for multivariate data in environmental science: <a href="https://cran.r-project.org/package=vegan">vegan</a> and <a href="https://cran.r-project.org/package=ade4">ade4</a>. Both packages partly overlap in their functionality. Here, you should choose whichever framework best suits your background.</p>
<ul>
<li><a href="https://cran.r-project.org/package=vegan">vegan</a>: ‘vegan’ is a community ecology package, which follows the approach of Mark Hill, Cajo ter Braak and others as well as the approach presented in <a href="https://www.elsevier.com/books/numerical-ecology/legendre/978-0-444-53868-0">Numerical Ecology</a> by Legendre &amp; Legendre (1988). It provides ordination methods, diversity analysis and other functions for community ecologists.</li>
<li><a href="https://cran.r-project.org/package=ade4">ade4</a>: Derives from the traditions of the French school of Analyse des Donnees and is based on the use of the duality diagram. The package provides tools for several multivariate analysis (i.e., ordination) of one-table (e.g., principal component analysis, correspondence analysis), two-table (e.g., coinertia analysis, redundancy analysis), three-table (e.g., RLQ analysis) and K-table (e.g., STATIS, multiple coinertia analysis) → see also <a href="https://www.jstatsoft.org/article/view/v022i04">Dray and Dufour (2007)</a>.</li>
</ul>
<p>Much ecological analysis proceeds from a matrix of dissimilarities between samples. Hence, a large amount of effort has been put into the development of a wide range of dissimilarity coefficients suitable for ecological data: <code>dist()</code> in the standard package ‘stats’, <code>daisy()</code> in ‘cluster’, <code>vegdist()</code> in ‘vegan’, <code>Dist()</code> in ‘amap’, <code>distance()</code> in ‘ecodist’, and a suite of functions in ade4.</p>
<p>See also the <a href="https://cran.r-project.org/web/views/Multivariate.html">Multivariate</a> CRAN task views for complementary information. The <a href="https://cran.r-project.org/web/views/Cluster.html">Cluster</a> task view provides a more detailed discussion of available cluster analysis methods and appropriate R functions and packages.</p>
<p>Other useful packages:</p>
<ul>
<li>The adehabitat family of packages is useful for the analysis of animal movements, home range and habitat selection by animals (<a href="https://cran.r-project.org/package=adehabitatLT">adehabitatLT</a>, <a href="https://cran.r-project.org/package=adehabitatHR">adehabitatHR</a>, <a href="https://cran.r-project.org/package=adehabitatHS">adehabitatHS</a>, and <a href="https://cran.r-project.org/package=adehabitatMA">adehabitatMA</a>)</li>
<li><a href="https://cran.r-project.org/package=amap">amap</a>: Another Multidimensional Analysis Package - Provides robust and parallelised methods such as a form of generalised and robust principal component analysis and alternative implementations of k-means and agglomerative hierarchical clustering.</li>
<li><a href="https://cran.r-project.org/package=biclust">biclust</a>: Provides several algorithms to find biclusters in two-dimensional data.</li>
<li><a href="https://CRAN.R-project.org/package=BiodiversityR">BiodivdersityR</a>: Package for Community Ecology and Suitability Analysis - Graphical User Interface (via the R-Commander) and utility functions (often based on the vegan package) for statistical analysis of biodiversity and ecological communities, including species accumulation curves, diversity indices, Renyi profiles, GLMs for analysis of species abundance and presence-absence, distance matrices, Mantel tests, and cluster, constrained and unconstrained ordination analysis. A book on biodiversity and community ecology analysis is available for free download from the <a href="http://www.worldagroforestry.org/output/tree-diversity-analysis">website</a>. In 2012, methods for (ensemble) suitability modelling and mapping were expanded in the package.</li>
<li><a href="https://cran.r-project.org/package=cluster">cluster</a>: cluster: “Finding Groups in Data”: Cluster Analysis Extended Rousseeuw et al. Methods for Cluster analysis. Much extended the original from Peter Rousseeuw, Anja Struyf and Mia Hubert, based on Kaufman and Rousseeuw (1990) “Finding Groups in Data”.</li>
<li><a href="https://cran.r-project.org/package=ecodist">ecodist</a>: Dissimilarity-based analysis functions including ordination and Mantel test functions, intended for use with spatial and community data.</li>
<li><a href="https://cran.r-project.org/package=FactoMineR">FactoMineR</a>: Provides a wide range of Factor Analysis methods, including multiple and hierarchical multiple factor analysis as well as multiple factor analysis of quantitative and qualitative data → see also the package <a href="http://factominer.free.fr">website</a>.</li>
<li><a href="https://cran.r-project.org/package=FD">FD</a>: sA package to compute different multidimensional functional diversity (FD) indices from multiple traits. It implements a distance-based framework to measure FD that allows any number and type of functional traits, and can also consider species relative abundances. It also contains other useful tools for functional ecology.</li>
<li><a href="https://cran.r-project.org/package=flexmix">flexmix</a>: The package provides implementations of model-based cluster analysis.</li>
<li><a href="https://cran.r-project.org/package=mda">mda</a>: For mixture and flexible discriminant analysis, multivariate adaptive regression splines (MARS), and adaptive spline backfitting (BRUTO).</li>
<li><a href="https://cran.r-project.org/package=mclust">mclust</a>: This package fits mixtures of Gaussians using the EM algorithm. It allows fine control of volume and shape of covariance matrices and agglomerative hierarchical clustering based on maximum likelihood. It provides comprehensive strategies using hierarchical clustering, EM and the Bayesian Information Criterion (BIC) for clustering, density estimation, and discriminant analysis → see also this <a href="https://cran.r-project.org/web/packages/mclust/vignettes/mclust.html">Quick Tour</a>.</li>
<li><a href="https://cran.r-project.org/package=mefa">mefa</a>: Provides functions for handling and reporting on multivariate count data in ecology and biogeography.</li>
<li><a href="https://cran.r-project.org/package=popbio">popbio</a>: Can be used to construct and analyse age- or stage-specific matrix population models. The package covers methods described in <a href="https://global.oup.com/ushe/product/matrix-population-models-9780878931217?cc=de&amp;lang=en&amp;">Matrix Population Models</a> by Caswell (2001) and <a href="https://global.oup.com/ushe/product/quantitative-conservation-biology-9780878935468?cc=de&amp;lang=en&amp;">Quantitative Conservation Biology</a> by Morris and Doak (2002).</li>
<li><a href="https://cran.r-project.org/package=primer">primer</a>: Functions and data for A Primer of Ecology with R - Primarily functions for systems of ordinary differential equations, difference equations, and eigenanalysis and projection of demographic matrices.</li>
<li><a href="https://cran.r-project.org/package=pvclust">pvclust</a>: A package for assessing the uncertainty in hierarchical cluster analysis. It provides approximately unbiased p-values as well as bootstrap p-values.</li>
</ul>
</div>
<div id="other-species-specific-packages" class="section level2">
<h2>Other species-specific packages</h2>
<ul>
<li><a href="https://cran.r-project.org/package=igraph">igraph</a>: Offers routines for simple graphs and network analysis. It can handle large graphs very well and provides functions for generating random and regular graphs, graph visualization, centrality methods and much more → see also <a href="https://igraph.org/r/" class="uri">https://igraph.org/r/</a>.</li>
<li><a href="https://cran.r-project.org/package=NetIndices">NetIndices</a>: Package for estimating network indices, including trophic structure of foodwebs in R. The following indices are included: Ascendency network indices, Direct and indirect dependencies, Effective measures, Environ network indices, General network indices, Pathway analysis, Network uncertainty indices and constraint efficiencies and the trophic level and omnivory indices of food webs.</li>
<li><a href="https://cran.r-project.org/web/packages/metacom/index.html">metacom</a>: Provides functions to analyze coherence, boundary clumping, and turnover following the pattern-based metacommunity analysis of <a href="https://dx.doi.org/10.1034/j.1600-0706.2002.970210.x">Leibold &amp; Mikkelson, 2002</a></li>
<li><a href="https://cran.r-project.org/web/packages/mizer/index.html">mizer</a>: A set of classes and methods to set up and run multi-species, trait based and community size spectrum ecological models, focused on the marine environment.</li>
<li><a href="https://cran.r-project.org/web/packages/rangeMapper/index.html">rangeMapper</a>: Provides tools for easy generation of (life-history) traits maps based on species range (extent-of-occurrence) maps.</li>
<li><a href="https://rpkgs.datanovia.com/survminer/index.html">survminer</a>: The ‘survminer’ R package provides functions for facilitating survival analysis and visualization such as survival curves, summary and diagnostics of Cox models → see also <a href="https://rpkgs.datanovia.com/survminer/index.html">here</a> for more.</li>
<li><a href="https://cran.r-project.org/web/packages/tRophicPosition/index.html">tRophicPosition</a>: – Bayesian estimation of trophic position from consumer stable isotope ratios.</li>
</ul>
</div>
<div id="text-analysis" class="section level2">
<h2>Text analysis</h2>
<ul>
<li><p><a href="https://CRAN.R-project.org/package=quanteda">quanteda</a>: quanteda: Quantitative Analysis of Textual Data A fast, flexible, and comprehensive framework for quantitative text analysis in R. Provides functionality for corpus management, creating and manipulating tokens and ngrams, exploring keywords in context, forming and manipulating sparse matrices of documents by features and feature co-occurrences, analyzing keywords, computing feature similarities and distances, applying content dictionaries, applying supervised and unsupervised machine learning, visually representing text and text analyses, and more. –&gt; see also this <a href="https://cran.r-project.org/web/packages/quanteda/vignettes/quickstart.html">Quick Start Guide</a>.</p></li>
<li><p><a href="https://CRAN.R-project.org/package=tidytext">tidytext</a>: Package for text mining for word processing and sentiment analysis using ‘dplyr’, ‘ggplot2’, and other tidy tools.</p></li>
</ul>
<!-- Footnotes: -->
</div>
</div>
<div class="footnotes">
<hr />
<ol>
<li id="fn1"><p>Jelovic, Dejan. “<a href="https://web.archive.org/web/20080211111923/http://www.jelovic.com/articles/why_java_is_slow.htm">Why Java Will Always Be Slower than C++</a>”. www.jelovic.com. Archived from the original on February 11, 2008. Retrieved November 19, 2018.<a href="#fnref1">↩</a></p></li>
<li id="fn2"><p><a href="https://en.wikipedia.org/wiki/Java_security" class="uri">https://en.wikipedia.org/wiki/Java_security</a><a href="#fnref2">↩</a></p></li>
</ol>
</div>

        </section>
      </div> 

      
      <div class="col-md-3" >
        <div id="toc">
          <h4 style="color:#6D6D70;">Overview</h4>
          <ul style="padding:10px;">
            <li><a href="#which-programming-languages-should-you-learn-for-marine-data-science">Programming languages to learn</a>
              <ul>
                <li><a href="#overview-of-programming-languages-and-applications">Overview of programming languages and applications</a></li>
              </ul>
            </li>
            <li><a href="#r-packages">R packages</a>
              <ul>
                <li><a href="#accessingretrieving-data">Accessing/retrieving data</a></li>
                <li><a href="#data-wrangling-and-processing">Data wrangling and processing</a></li>
                <li><a href="#general-visualisation-packages">General visualisation packages</a></li>
                <li><a href="#time-series">Time series</a></li>
                <li><a href="#spatial-analysis">Spatial analysis</a></li>
                <li><a href="#oceanographic-analysis">Oceanographic analysis</a></li>
                <li><a href="#fishy-packages-for-analysis">Fishy packages for analysis</a></li>
                <li><a href="#modelling-species-responses-and-other-data-supervised-learning">Modelling species responses and other data (supervised learning)</a></li>
                <li><a href="#multivariate-analyses-supervised-and-unsupervised">Multivariate analyses (supervised and unsupervised)</a></li>
                <li><a href="#other-species-specific-packages">Other species-specific packages</a></li>
                <li><a href="#text-analysis">Text analysis</a></li>
              </ul>
            </li>
          </ul>
        </div>
      </div>

  </div>
</div>





            
        </div>
        

        <footer id="footer">
    <div class="container">

        
        <div class="col-md-4 col-sm-6">
            <h4>About me</h4>
            <img src="/img/profile_v1@2x.png" alt="profile" style="float:left;width:75px;height:75px;padding:5px;"/>

            I am a quantitative marine ecologist who loves combining nerdy programming, statistics, and any activity in and around the oceans.

            <hr class="hidden-md hidden-lg hidden-sm">

        </div>
        
        

        <div class="col-md-4 col-sm-6">

             
            <h4>Recent posts</h4>

            <div class="blog-entries">
                
                <div class="item same-height-row clearfix">
                    <div class="image same-height-always">
                        <a href="/blog/2018/10/17/r-code-test/">
                          
                            <img src="/img/placeholder.png" class="img-responsive" alt="R code test">
                          
                        </a>
                    </div>
                    <div class="name same-height-always">
                        <h5><a href="/blog/2018/10/17/r-code-test/">R code test</a></h5>
                    </div>
                </div>
                
                <div class="item same-height-row clearfix">
                    <div class="image same-height-always">
                        <a href="/blog/2018/10/16/testing-indicator-performance/">
                          
                            <img src="/img/banners/indicators.jpg" class="img-responsive" alt="Testing indicator performance">
                          
                        </a>
                    </div>
                    <div class="name same-height-always">
                        <h5><a href="/blog/2018/10/16/testing-indicator-performance/">Testing indicator performance</a></h5>
                    </div>
                </div>
                
                <div class="item same-height-row clearfix">
                    <div class="image same-height-always">
                        <a href="/blog/">
                          
                            <img src="/img/placeholder.png" class="img-responsive" alt="Blogs">
                          
                        </a>
                    </div>
                    <div class="name same-height-always">
                        <h5><a href="/blog/">Blogs</a></h5>
                    </div>
                </div>
                
            </div>

            <hr class="hidden-md hidden-lg">
             

        </div>
        

        
        <div class="col-md-4 col-sm-6">

          <h4>Contact</h4>

            <strong><a href="https://www.biologie.uni-hamburg.de/en/einrichtungen/imf.html">Institute of Marine Ecosystem and Fishery Science</a></strong>
        <br>Center for Earth System Research and Sustainability
        <br>University of Hamburg
        <br><strong>Germany</strong><br>
      

            <a href="/contact" class="btn btn-small btn-template-main">Go to contact page</a>

            <hr class="hidden-md hidden-lg hidden-sm">

        </div>
        
        

    </div>
    
</footer>







<div id="copyright">
    <div class="container">
        <div class="col-md-12">
            
            <p class="pull-left">This work is licensed under a <a rel='license' href='http://creativecommons.org/licenses/by-sa/4.0/'>Creative Commons Attribution-ShareAlike 4.0 International License </a></p>
            
            <p class="pull-right">
              Template by <a href="http://bootstrapious.com/free-templates">Bootstrapious</a>.
              

              Ported to Hugo by <a href="https://github.com/devcows/hugo-universal-theme">DevCows</a>
            </p>
        </div>
    </div>
</div>





    </div>
    

    
<script src="//code.jquery.com/jquery-3.1.1.min.js" integrity="sha256-hVVnYaiADRTO2PzUGmuLJr8BLUSjGIZsDYGmIJLv2b8=" crossorigin="anonymous"></script>
<script src="//maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<script src="//cdnjs.cloudflare.com/ajax/libs/jquery-cookie/1.4.1/jquery.cookie.min.js"></script>
<script src="//cdnjs.cloudflare.com/ajax/libs/waypoints/4.0.1/jquery.waypoints.min.js"></script>
<script src="//cdnjs.cloudflare.com/ajax/libs/Counter-Up/1.0/jquery.counterup.min.js"></script>
<script src="//cdnjs.cloudflare.com/ajax/libs/jquery-parallax/1.1.3/jquery-parallax.js"></script>

<script src="//maps.googleapis.com/maps/api/js?v=3.exp"></script>

<script src="/js/hpneo.gmaps.js"></script>
<script src="/js/gmaps.init.js"></script>
<script src="/js/front.js"></script>


<script src="/js/owl.carousel.min.js"></script>



<script>
  var shiftWindow = function() { scrollBy(0, -200) };
  if (location.hash) shiftWindow();
  window.addEventListener("hashchange", shiftWindow);
</script>



<script>

function make_sticky(id) {
    var e = $(id);
    var w = $(window);
    $('<div/>').insertBefore(id);
    $('<div/>').hide().css('height',e.outerHeight()).insertAfter(id);
    var n = e.next();
    var p = e.prev();
    function sticky_relocate() {
      var window_top = w.scrollTop();
      var div_top = p.offset().top;
      if (window_top > div_top) {
        e.addClass('sticky');
        n.show();
      } else {
        e.removeClass('sticky');
        n.hide();
      }
    }
    w.scroll(sticky_relocate);
    sticky_relocate();
}
make_sticky('#toc');
</script>






  </body>
</html>
